<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <meta name="author" content="Shing Wan CHOI">
  
  <link rel="shortcut icon" href="../img/favicon.ico">
  <title>LDpred-2 - Basic Tutorial for Polygenic Risk Score Analyses</title>
  <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Lato:400,700|Roboto+Slab:400,700|Inconsolata:400,700" />

  <link rel="stylesheet" href="../css/theme.css" />
  <link rel="stylesheet" href="../css/theme_extra.css" />
  <link href="../css/extra.css" rel="stylesheet" />
  <link href="../css/details.css" rel="stylesheet" />
  
  <script>
    // Current page data
    var mkdocs_page_name = "LDpred-2";
    var mkdocs_page_input_path = "ldpred.md";
    var mkdocs_page_url = null;
  </script>
  
  <script src="../js/jquery-2.1.1.min.js" defer></script>
  <script src="../js/modernizr-2.8.3.min.js" defer></script> 
  
  <script>
      (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
      (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
      m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
      })(window,document,'script','https://www.google-analytics.com/analytics.js','ga');

      ga('create', 'UA-145068952-1', 'auto');
      ga('send', 'pageview');
  </script>
  
</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side stickynav">
    <div class="wy-side-scroll">
      <div class="wy-side-nav-search">
        <a href=".." class="icon icon-home"> Basic Tutorial for Polygenic Risk Score Analyses</a>
        <div role="search">
  <form id ="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" title="Type search term here" />
  </form>
</div>
      </div>

      <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
                <ul>
                    <li class="toctree-l1"><a class="reference internal" href="..">Home</a>
                    </li>
                </ul>
                <ul>
                    <li class="toctree-l1"><a class="reference internal" href="../base/">1. QC of Base Data</a>
                    </li>
                </ul>
                <ul>
                    <li class="toctree-l1"><a class="reference internal" href="../target/">2. QC of Target Data</a>
                    </li>
                </ul>
                <ul>
                    <li class="toctree-l1"><a class="reference internal" href="../cal_prs/">3. Calculating and analysing PRS</a>
                    </li>
                </ul>
                <ul>
                    <li class="toctree-l1"><a class="reference internal" href="../plink/">PLINK</a>
                    </li>
                </ul>
                <ul>
                    <li class="toctree-l1"><a class="reference internal" href="../prsice/">PRSice-2</a>
                    </li>
                </ul>
                <ul class="current">
                    <li class="toctree-l1 current"><a class="reference internal current" href="./">LDpred-2</a>
    <ul class="current">
    <li class="toctree-l2"><a class="reference internal" href="#installing-ldpred-2">Installing LDpred-2</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#required-data">Required Data</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#running-prs-analysis">Running PRS analysis</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#0-prepare-workspace">0. Prepare workspace</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#1-read-in-the-phenotype-and-covariate-files">1. Read in the phenotype and covariate files</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#2-load-in-the-genotype-and-summary-statistic-file-using-bigsnpr">2. Load in the genotype and summary statistic file using bigsnpr</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#3-reformat-data-in-preparation-of-ldpred2">3. Reformat data in preparation of LDpred2</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#4-start-running-ldpred-2">4. Start running LDpred 2</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#5-get-the-final-performance-of-the-ldpred-models">5. Get the final performance of the LDpred models</a>
    </li>
        </ul>
    </li>
    </ul>
                    </li>
                </ul>
                <ul>
                    <li class="toctree-l1"><a class="reference internal" href="../lassosum/">lassosum</a>
                    </li>
                </ul>
                <ul>
                    <li class="toctree-l1"><a class="reference internal" href="../plink_visual/">4. Visualizing PRS Results</a>
                    </li>
                </ul>
      </div>
    </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" role="navigation" aria-label="top navigation">
        <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
        <a href="..">Basic Tutorial for Polygenic Risk Score Analyses</a>
      </nav>

      
      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href="..">Docs</a> &raquo;</li>
    
      
    
    <li>LDpred-2</li>
    <li class="wy-breadcrumbs-aside">
      
        <a href="https://github.com/choishingwan/PRS-Tutorial/edit/master/docs/ldpred.md"
          class="icon icon-github"> Edit on GitHub</a>
      
    </li>
  </ul>
  
  <hr/>
</div>
          <div role="main">
            <div class="section">
              
                <h1 id="background">Background<a class="headerlink" href="#background" title="Permanent link">&para;</a></h1>
<p><a href="https://privefl.github.io/bigsnpr/articles/LDpred2.html">LDpred-2</a> is one of the dedicated PRS programs which is an <code>R</code> package that uses a Bayesian approach to polygenic risk scoring.</p>
<h2 id="installing-ldpred-2">Installing LDpred-2<a class="headerlink" href="#installing-ldpred-2" title="Permanent link">&para;</a></h2>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The script used here is based on LDpred 2 implemented under bigsnpr version 1.4.7</p>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>For more details, please refer to <a href="https://privefl.github.io/bigsnpr/articles/LDpred2.html">LDpred 2's homepage</a></p>
</div>
<p>You can install <code>LDpred</code> and its dependencies in <code>R</code> with the following command:</p>
<div class="highlight" style="background: #f8f8f8"><pre style="line-height: 125%"><span></span><code><span style="color: #0000FF">install.packages</span>(<span style="color: #BA2121">&quot;remotes&quot;</span>)
<span style="color: #0000FF">library</span>(remotes)
remotes<span style="color: #666666">::</span><span style="color: #0000FF">install_github</span>(<span style="color: #BA2121">&quot;https://github.com/privefl/bigsnpr.git&quot;</span>)
</code></pre></div>

<div class="admonition note">
<p class="admonition-title">Note</p>
<p>For mac users, you might need to follow the guide <a href="https://thecoatlessprofessor.com/programming/cpp/openmp-in-r-on-os-x/">here</a> to be able to install LDpred2</p>
</div>
<h2 id="required-data">Required Data<a class="headerlink" href="#required-data" title="Permanent link">&para;</a></h2>
<p>We assume that you have the following files (or you can download it from <a href="https://drive.google.com/file/d/1x_G0Gxk9jFMY-PMqwtg6-vdEyUPp5p5u/view?usp=sharing">here</a>):</p>
<table>
<thead>
<tr>
<th align="center">File Name</th>
<th align="center">Description</th>
</tr>
</thead>
<tbody>
<tr>
<td align="center"><strong>Height.QC.gz</strong></td>
<td align="center">The post-QCed summary statistic</td>
</tr>
<tr>
<td align="center"><strong>EUR.QC.bed</strong></td>
<td align="center">The genotype file after performing some basic filtering</td>
</tr>
<tr>
<td align="center"><strong>EUR.QC.bim</strong></td>
<td align="center">This file contains the SNPs that passed the basic filtering</td>
</tr>
<tr>
<td align="center"><strong>EUR.QC.fam</strong></td>
<td align="center">This file contains the samples that passed the basic filtering</td>
</tr>
<tr>
<td align="center"><strong>EUR.height</strong></td>
<td align="center">This file contains the phenotype of the samples</td>
</tr>
<tr>
<td align="center"><strong>EUR.cov</strong></td>
<td align="center">This file contains the covariates of the samples</td>
</tr>
<tr>
<td align="center"><strong>EUR.eigenvec</strong></td>
<td align="center">This file contains the PCs of the samples</td>
</tr>
</tbody>
</table>
<h2 id="running-prs-analysis">Running PRS analysis<a class="headerlink" href="#running-prs-analysis" title="Permanent link">&para;</a></h2>
<h3 id="0-prepare-workspace">0. Prepare workspace<a class="headerlink" href="#0-prepare-workspace" title="Permanent link">&para;</a></h3>
<p>On some server, you might need to first use the following code in order to run LDpred with multi-thread</p>
<div class="tabbed-set" data-tabs="1:1"><input checked="checked" id="__tabbed_1_1" name="__tabbed_1" type="radio" /><label for="__tabbed_1_1">prepare workspace</label><div class="tabbed-content">
<div class="highlight" style="background: #f8f8f8"><pre style="line-height: 125%"><span></span><code><span style="color: #0000FF">options</span>(bigstatsr.check.parallel.blas <span style="color: #666666">=</span> <span style="color: #008000; font-weight: bold">FALSE</span>)
<span style="color: #0000FF">options</span>(default.nproc.blas <span style="color: #666666">=</span> <span style="color: #008000; font-weight: bold">NULL</span>)
</code></pre></div>

</div>
</div>
<h3 id="1-read-in-the-phenotype-and-covariate-files">1. Read in the phenotype and covariate files<a class="headerlink" href="#1-read-in-the-phenotype-and-covariate-files" title="Permanent link">&para;</a></h3>
<div class="tabbed-set" data-tabs="2:1"><input checked="checked" id="__tabbed_2_1" name="__tabbed_2" type="radio" /><label for="__tabbed_2_1">read in phenotype and covariates</label><div class="tabbed-content">
<div class="highlight" style="background: #f8f8f8"><pre style="line-height: 125%"><span></span><code><span style="color: #0000FF">library</span>(data.table)
<span style="color: #0000FF">library</span>(magrittr)
phenotype <span style="color: #666666">&lt;-</span> <span style="color: #0000FF">fread</span>(<span style="color: #BA2121">&quot;EUR.height&quot;</span>)
covariate <span style="color: #666666">&lt;-</span> <span style="color: #0000FF">fread</span>(<span style="color: #BA2121">&quot;EUR.cov&quot;</span>)
pcs <span style="color: #666666">&lt;-</span> <span style="color: #0000FF">fread</span>(<span style="color: #BA2121">&quot;EUR.eigenvec&quot;</span>)
<span style="color: #408080; font-style: italic"># rename columns</span>
<span style="color: #0000FF">colnames</span>(pcs) <span style="color: #666666">&lt;-</span> <span style="color: #0000FF">c</span>(<span style="color: #BA2121">&quot;FID&quot;</span>,<span style="color: #BA2121">&quot;IID&quot;</span>, <span style="color: #0000FF">paste0</span>(<span style="color: #BA2121">&quot;PC&quot;</span>,<span style="color: #666666">1:6</span>))
<span style="color: #408080; font-style: italic"># generate required table</span>
pheno <span style="color: #666666">&lt;-</span> <span style="color: #0000FF">merge</span>(phenotype, covariate) <span style="color: #666666">%&gt;%</span>
    <span style="color: #0000FF">merge</span>(., pcs)
</code></pre></div>

</div>
</div>
<h3 id="2-load-in-the-genotype-and-summary-statistic-file-using-bigsnpr">2. Load in the genotype and summary statistic file using bigsnpr<a class="headerlink" href="#2-load-in-the-genotype-and-summary-statistic-file-using-bigsnpr" title="Permanent link">&para;</a></h3>
<div class="tabbed-set" data-tabs="3:1"><input checked="checked" id="__tabbed_3_1" name="__tabbed_3" type="radio" /><label for="__tabbed_3_1">Load genotype with bigsnpr</label><div class="tabbed-content">
<div class="highlight" style="background: #f8f8f8"><pre style="line-height: 125%"><span></span><code><span style="color: #408080; font-style: italic"># load the library</span>
<span style="color: #0000FF">library</span>(bigsnpr)
<span style="color: #408080; font-style: italic"># now preprocess the bed file. Only need to do once for each data set</span>
<span style="color: #0000FF">snp_readBed</span>(<span style="color: #BA2121">&quot;EUR.QC.bed&quot;</span>)
<span style="color: #408080; font-style: italic"># now attach the genotype object</span>
obj.bigSNP <span style="color: #666666">&lt;-</span> <span style="color: #0000FF">snp_attach</span>(<span style="color: #BA2121">&quot;EUR.QC.rds&quot;</span>)
<span style="color: #408080; font-style: italic"># Next, read in the summary statistic file</span>
sumstats <span style="color: #666666">&lt;-</span> bigreadr<span style="color: #666666">::</span><span style="color: #0000FF">fread2</span>(<span style="color: #BA2121">&quot;Height.QC.gz&quot;</span>)
</code></pre></div>

</div>
</div>
<h3 id="3-reformat-data-in-preparation-of-ldpred2">3. Reformat data in preparation of LDpred2<a class="headerlink" href="#3-reformat-data-in-preparation-of-ldpred2" title="Permanent link">&para;</a></h3>
<div class="tabbed-set" data-tabs="4:1"><input checked="checked" id="__tabbed_4_1" name="__tabbed_4" type="radio" /><label for="__tabbed_4_1">data reformating</label><div class="tabbed-content">
<div class="highlight" style="background: #f8f8f8"><pre style="line-height: 125%"><span></span><code><span style="color: #408080; font-style: italic"># LDpred 2 require the header to follow the exact naming</span>
<span style="color: #0000FF">names</span>(sumstats) <span style="color: #666666">&lt;-</span>
    <span style="color: #0000FF">c</span>(<span style="color: #BA2121">&quot;chr&quot;</span>,
    <span style="color: #BA2121">&quot;pos&quot;</span>,
    <span style="color: #BA2121">&quot;rsid&quot;</span>,
    <span style="color: #BA2121">&quot;a1&quot;</span>,
    <span style="color: #BA2121">&quot;a0&quot;</span>,
    <span style="color: #BA2121">&quot;n_eff&quot;</span>,
    <span style="color: #BA2121">&quot;beta_se&quot;</span>,
    <span style="color: #BA2121">&quot;p&quot;</span>,
    <span style="color: #BA2121">&quot;OR&quot;</span>,
    <span style="color: #BA2121">&quot;INFO&quot;</span>,
    <span style="color: #BA2121">&quot;MAF&quot;</span>)
<span style="color: #408080; font-style: italic"># Transform the OR into log(OR)</span>
sumstats<span style="color: #666666">$</span>beta <span style="color: #666666">&lt;-</span> <span style="color: #0000FF">log</span>(sumstats<span style="color: #666666">$</span>OR)
<span style="color: #408080; font-style: italic"># extract the SNP information from the genotype</span>
map <span style="color: #666666">&lt;-</span> obj.bigSNP<span style="color: #666666">$</span>map[<span style="color: #666666">-2</span>]
<span style="color: #0000FF">names</span>(map) <span style="color: #666666">&lt;-</span> <span style="color: #0000FF">c</span>(<span style="color: #BA2121">&quot;chr&quot;</span>, <span style="color: #BA2121">&quot;rsid&quot;</span>, <span style="color: #BA2121">&quot;pos&quot;</span>, <span style="color: #BA2121">&quot;a1&quot;</span>, <span style="color: #BA2121">&quot;a0&quot;</span>)
<span style="color: #408080; font-style: italic"># perform SNP matching</span>
info_snp <span style="color: #666666">&lt;-</span> <span style="color: #0000FF">snp_match</span>(sumstats, map)
<span style="color: #408080; font-style: italic"># Rename the data structures</span>
CHR <span style="color: #666666">&lt;-</span> map<span style="color: #666666">$</span>chr
POS <span style="color: #666666">&lt;-</span> map<span style="color: #666666">$</span>pos
<span style="color: #408080; font-style: italic"># get the CM information from 1000 Genome</span>
<span style="color: #408080; font-style: italic"># will download the 1000G file to the current directory (&quot;.&quot;)</span>
POS2 <span style="color: #666666">&lt;-</span> <span style="color: #0000FF">snp_asGeneticPos</span>(CHR, POS, dir <span style="color: #666666">=</span> <span style="color: #BA2121">&quot;.&quot;</span>)
<span style="color: #408080; font-style: italic"># Reformat the phenotype information to match the fam order</span>
fam.order <span style="color: #666666">&lt;-</span> <span style="color: #0000FF">as.data.table</span>(obj.bigSNP<span style="color: #666666">$</span>fam)
<span style="color: #0000FF">setnames</span>(fam.order,
        <span style="color: #0000FF">c</span>(<span style="color: #BA2121">&quot;family.ID&quot;</span>, <span style="color: #BA2121">&quot;sample.ID&quot;</span>),
        <span style="color: #0000FF">c</span>(<span style="color: #BA2121">&quot;FID&quot;</span>, <span style="color: #BA2121">&quot;IID&quot;</span>))
<span style="color: #408080; font-style: italic"># This will ensure our phenotype vector y is of the same order as the</span>
<span style="color: #408080; font-style: italic"># family object generated by bigsnpr</span>
y <span style="color: #666666">&lt;-</span> pheno[fam.order, on <span style="color: #666666">=</span> <span style="color: #0000FF">c</span>(<span style="color: #BA2121">&quot;FID&quot;</span>, <span style="color: #BA2121">&quot;IID&quot;</span>)]
genotype <span style="color: #666666">&lt;-</span> obj.bigSNP<span style="color: #666666">$</span>genotypes
<span style="color: #408080; font-style: italic"># Use all sample for test</span>
ind.test <span style="color: #666666">&lt;-</span> <span style="color: #666666">1:</span><span style="color: #0000FF">nrow</span>(genotype)
<span style="color: #408080; font-style: italic"># Calculate the null R2</span>
<span style="color: #408080; font-style: italic"># use glm for binary trait </span>
<span style="color: #408080; font-style: italic"># (will also need the fmsb package to calculate the pseudo R2)</span>
null.model <span style="color: #666666">&lt;-</span> <span style="color: #0000FF">paste</span>(<span style="color: #BA2121">&quot;PC&quot;</span>, <span style="color: #666666">1:6</span>, sep <span style="color: #666666">=</span> <span style="color: #BA2121">&quot;&quot;</span>, collapse <span style="color: #666666">=</span> <span style="color: #BA2121">&quot;+&quot;</span>) <span style="color: #666666">%&gt;%</span>
    <span style="color: #0000FF">paste0</span>(<span style="color: #BA2121">&quot;Height~Sex+&quot;</span>, .) <span style="color: #666666">%&gt;%</span>
    as.formula <span style="color: #666666">%&gt;%</span>
    <span style="color: #0000FF">lm</span>(., data <span style="color: #666666">=</span> y) <span style="color: #666666">%&gt;%</span>
    summary
null.r2 <span style="color: #666666">&lt;-</span> null.model<span style="color: #666666">$</span>r.squared
reg.formula <span style="color: #666666">&lt;-</span> <span style="color: #0000FF">paste</span>(<span style="color: #BA2121">&quot;PC&quot;</span>, <span style="color: #666666">1:6</span>, sep <span style="color: #666666">=</span> <span style="color: #BA2121">&quot;&quot;</span>, collapse <span style="color: #666666">=</span> <span style="color: #BA2121">&quot;+&quot;</span>) <span style="color: #666666">%&gt;%</span>
    <span style="color: #0000FF">paste0</span>(<span style="color: #BA2121">&quot;Height~PRS+Sex+&quot;</span>, .) <span style="color: #666666">%&gt;%</span>
    as.formula
</code></pre></div>

</div>
</div>
<h3 id="4-start-running-ldpred-2">4. Start running LDpred 2<a class="headerlink" href="#4-start-running-ldpred-2" title="Permanent link">&para;</a></h3>
<div class="tabbed-set" data-tabs="5:4"><input checked="checked" id="__tabbed_5_1" name="__tabbed_5" type="radio" /><label for="__tabbed_5_1">infinitesimal model</label><div class="tabbed-content">
<div class="highlight" style="background: #f8f8f8"><pre style="line-height: 125%"><span></span><code><span style="color: #408080; font-style: italic"># Get maximum amount of cores</span>
NCORES <span style="color: #666666">&lt;-</span> <span style="color: #0000FF">nb_cores</span>()
<span style="color: #408080; font-style: italic"># Start doing analysis on each chromosome</span>
y[, Inf.est <span style="color: #666666">:=</span> <span style="color: #666666">0</span>]
<span style="color: #408080; font-style: italic"># add progress bar</span>
pb <span style="color: #666666">=</span> <span style="color: #0000FF">txtProgressBar</span>(min <span style="color: #666666">=</span> <span style="color: #666666">0</span>, max <span style="color: #666666">=</span> <span style="color: #666666">22</span>, initial <span style="color: #666666">=</span> <span style="color: #666666">0</span>)
<span style="color: #0000FF">for </span>(chr in <span style="color: #666666">1:22</span>) {
    <span style="color: #0000FF">setTxtProgressBar</span>(pb, chr)
    <span style="color: #408080; font-style: italic"># extract current chromosome</span>
    chr.idx <span style="color: #666666">&lt;-</span> <span style="color: #0000FF">which</span>(info_snp<span style="color: #666666">$</span>chr <span style="color: #666666">==</span> chr)
    df_beta <span style="color: #666666">&lt;-</span> info_snp[chr.idx,
                        <span style="color: #0000FF">c</span>(<span style="color: #BA2121">&quot;beta&quot;</span>, <span style="color: #BA2121">&quot;beta_se&quot;</span>, <span style="color: #BA2121">&quot;n_eff&quot;</span>)]
    ind.chr <span style="color: #666666">&lt;-</span> info_snp<span style="color: #666666">$</span>`_NUM_ID_`[chr.idx]
    <span style="color: #408080; font-style: italic"># calculate LD</span>
    corr0 <span style="color: #666666">&lt;-</span> <span style="color: #0000FF">snp_cor</span>(
        genotype,
        ind.col <span style="color: #666666">=</span> ind.chr,
        ncores <span style="color: #666666">=</span> NCORES,
        infos.pos <span style="color: #666666">=</span> POS2[ind.chr],
        size <span style="color: #666666">=</span> <span style="color: #666666">3</span> <span style="color: #666666">/</span> <span style="color: #666666">1000</span>
    )
    corr <span style="color: #666666">&lt;-</span> bigsparser<span style="color: #666666">::</span><span style="color: #0000FF">as_SFBM</span>(<span style="color: #0000FF">as</span>(corr0, <span style="color: #BA2121">&quot;dgCMatrix&quot;</span>))
    <span style="color: #408080; font-style: italic"># Perform LDSC analysis to get h2 estimate</span>
    ldsc <span style="color: #666666">&lt;-</span> <span style="color: #0000FF">snp_ldsc2</span>(corr0, df_beta)
    h2_est <span style="color: #666666">&lt;-</span> ldsc[[<span style="color: #BA2121">&quot;h2&quot;</span>]]
    <span style="color: #408080; font-style: italic"># Get adjusted beta from infinitesimal model</span>
    beta_inf <span style="color: #666666">&lt;-</span> <span style="color: #0000FF">snp_ldpred2_inf</span>(corr, df_beta, h2 <span style="color: #666666">=</span> h2_est)
    <span style="color: #408080; font-style: italic"># Get infinitesimal PRS</span>
    pred_inf <span style="color: #666666">&lt;-</span> <span style="color: #0000FF">big_prodVec</span>(genotype,
                            beta_inf,
                            ind.row <span style="color: #666666">=</span> ind.test,
                            ind.col <span style="color: #666666">=</span> ind.chr)
    <span style="color: #408080; font-style: italic"># add up the calculated PRS</span>
    y[, Inf.est <span style="color: #666666">:=</span> Inf.est <span style="color: #666666">+</span> pred_inf]
}
<span style="color: #0000FF">print</span>(<span style="color: #BA2121">&quot;Completed&quot;</span>)
</code></pre></div>

</div>
<input id="__tabbed_5_2" name="__tabbed_5" type="radio" /><label for="__tabbed_5_2">grid model</label><div class="tabbed-content">
<div class="highlight" style="background: #f8f8f8"><pre style="line-height: 125%"><span></span><code><span style="color: #408080; font-style: italic"># Get maximum amount of cores</span>
NCORES <span style="color: #666666">&lt;-</span> <span style="color: #0000FF">nb_cores</span>()
<span style="color: #408080; font-style: italic"># Prepare data for grid model</span>
p_seq <span style="color: #666666">&lt;-</span> <span style="color: #0000FF">signif</span>(<span style="color: #0000FF">seq_log</span>(<span style="color: #666666">1e-4</span>, <span style="color: #666666">1</span>, length.out <span style="color: #666666">=</span> <span style="color: #666666">17</span>), <span style="color: #666666">2</span>)
<span style="color: #408080; font-style: italic"># Start doing analysis on each chromosome</span>
y[, grid <span style="color: #666666">:=</span> <span style="color: #666666">0</span>]
<span style="color: #408080; font-style: italic"># add progress bar</span>
pb <span style="color: #666666">=</span> <span style="color: #0000FF">txtProgressBar</span>(min <span style="color: #666666">=</span> <span style="color: #666666">0</span>, max <span style="color: #666666">=</span> <span style="color: #666666">22</span>, initial <span style="color: #666666">=</span> <span style="color: #666666">0</span>)
<span style="color: #0000FF">for </span>(chr in <span style="color: #666666">1:22</span>) {
    <span style="color: #0000FF">setTxtProgressBar</span>(pb, chr)
    <span style="color: #408080; font-style: italic"># extract current chromosome</span>
    chr.idx <span style="color: #666666">&lt;-</span> <span style="color: #0000FF">which</span>(info_snp<span style="color: #666666">$</span>chr <span style="color: #666666">==</span> chr)
    df_beta <span style="color: #666666">&lt;-</span> info_snp[chr.idx,
                        <span style="color: #0000FF">c</span>(<span style="color: #BA2121">&quot;beta&quot;</span>, <span style="color: #BA2121">&quot;beta_se&quot;</span>, <span style="color: #BA2121">&quot;n_eff&quot;</span>)]
    ind.chr <span style="color: #666666">&lt;-</span> info_snp<span style="color: #666666">$</span>`_NUM_ID_`[chr.idx]
    <span style="color: #408080; font-style: italic"># calculate LD</span>
    corr0 <span style="color: #666666">&lt;-</span> <span style="color: #0000FF">snp_cor</span>(
        genotype,
        ind.col <span style="color: #666666">=</span> ind.chr,
        ncores <span style="color: #666666">=</span> NCORES,
        infos.pos <span style="color: #666666">=</span> POS2[ind.chr],
        size <span style="color: #666666">=</span> <span style="color: #666666">3</span> <span style="color: #666666">/</span> <span style="color: #666666">1000</span>
    )
    corr <span style="color: #666666">&lt;-</span> bigsparser<span style="color: #666666">::</span><span style="color: #0000FF">as_SFBM</span>(<span style="color: #0000FF">as</span>(corr0, <span style="color: #BA2121">&quot;dgCMatrix&quot;</span>))
    <span style="color: #408080; font-style: italic"># Perform LDSC analysis to get h2 estimate</span>
    ldsc <span style="color: #666666">&lt;-</span> <span style="color: #0000FF">snp_ldsc2</span>(corr0, df_beta)
    h2_est <span style="color: #666666">&lt;-</span> ldsc[[<span style="color: #BA2121">&quot;h2&quot;</span>]]
    h2_seq <span style="color: #666666">&lt;-</span> <span style="color: #0000FF">round</span>(h2_est <span style="color: #666666">*</span> <span style="color: #0000FF">c</span>(<span style="color: #666666">0.7</span>, <span style="color: #666666">1</span>, <span style="color: #666666">1.4</span>), <span style="color: #666666">4</span>)
    grid.param <span style="color: #666666">&lt;-</span>
        <span style="color: #0000FF">expand.grid</span>(p <span style="color: #666666">=</span> p_seq,
                    h2 <span style="color: #666666">=</span> h2_seq,
                    sparse <span style="color: #666666">=</span> <span style="color: #0000FF">c</span>(<span style="color: #008000; font-weight: bold">FALSE</span>, <span style="color: #008000; font-weight: bold">TRUE</span>))
    <span style="color: #408080; font-style: italic"># Get adjusted beta from grid model</span>
    beta_grid <span style="color: #666666">&lt;-</span>
        <span style="color: #0000FF">snp_ldpred2_grid</span>(corr, df_beta, grid.param, ncores <span style="color: #666666">=</span> NCORES)
    <span style="color: #408080; font-style: italic"># Get the grid PRSs</span>
    pred_grid <span style="color: #666666">&lt;-</span> <span style="color: #0000FF">big_prodMat</span>(genotype, beta_grid, ind.col <span style="color: #666666">=</span> ind.chr)
    <span style="color: #408080; font-style: italic"># Find the current &quot;best&quot; PRS</span>
    indep <span style="color: #666666">&lt;-</span> <span style="color: #0000FF">copy</span>(y)
    max.grid.r2 <span style="color: #666666">&lt;-</span> <span style="color: #666666">0</span>
    max.grid.id <span style="color: #666666">&lt;-</span> <span style="color: #666666">0</span>
    <span style="color: #0000FF">for </span>(i in <span style="color: #666666">1:</span><span style="color: #0000FF">ncol</span>(pred_grid)) {
        indep[, PRS <span style="color: #666666">:=</span> pred_grid[, i]]
        model <span style="color: #666666">&lt;-</span> <span style="color: #0000FF">lm</span>(reg.formula, data <span style="color: #666666">=</span> indep) <span style="color: #666666">%&gt;%</span>
            summary
        prs.r2 <span style="color: #666666">&lt;-</span> model<span style="color: #666666">$</span>r.squared <span style="color: #666666">-</span> null.r2
        <span style="color: #0000FF">if </span>(max.grid.r2 <span style="color: #666666">&lt;</span> prs.r2) {
            max.grid.r2 <span style="color: #666666">&lt;-</span> prs.r2
            max.grid.id <span style="color: #666666">&lt;-</span> i
        }
    }
    <span style="color: #408080; font-style: italic"># add up the calculated grid PRS</span>
    y[, grid <span style="color: #666666">:=</span> grid <span style="color: #666666">+</span> pred_grid[, max.grid.id]]
}
<span style="color: #0000FF">print</span>(<span style="color: #BA2121">&quot;Completed&quot;</span>)
</code></pre></div>

</div>
<input id="__tabbed_5_3" name="__tabbed_5" type="radio" /><label for="__tabbed_5_3">auto model</label><div class="tabbed-content">
<div class="highlight" style="background: #f8f8f8"><pre style="line-height: 125%"><span></span><code><span style="color: #408080; font-style: italic"># Get maximum amount of cores</span>
NCORES <span style="color: #666666">&lt;-</span> <span style="color: #0000FF">nb_cores</span>()
<span style="color: #408080; font-style: italic"># Start doing analysis on each chromosome</span>
y[, auto <span style="color: #666666">:=</span> <span style="color: #666666">0</span>]
<span style="color: #408080; font-style: italic"># add progress bar</span>
pb <span style="color: #666666">=</span> <span style="color: #0000FF">txtProgressBar</span>(min <span style="color: #666666">=</span> <span style="color: #666666">0</span>, max <span style="color: #666666">=</span> <span style="color: #666666">22</span>, initial <span style="color: #666666">=</span> <span style="color: #666666">0</span>)
<span style="color: #0000FF">for </span>(chr in <span style="color: #666666">1:22</span>) {
    <span style="color: #0000FF">setTxtProgressBar</span>(pb, chr)
    <span style="color: #408080; font-style: italic"># extract current chromosome</span>
    chr.idx <span style="color: #666666">&lt;-</span> <span style="color: #0000FF">which</span>(info_snp<span style="color: #666666">$</span>chr <span style="color: #666666">==</span> chr)
    df_beta <span style="color: #666666">&lt;-</span> info_snp[chr.idx,
                        <span style="color: #0000FF">c</span>(<span style="color: #BA2121">&quot;beta&quot;</span>, <span style="color: #BA2121">&quot;beta_se&quot;</span>, <span style="color: #BA2121">&quot;n_eff&quot;</span>)]
    ind.chr <span style="color: #666666">&lt;-</span> info_snp<span style="color: #666666">$</span>`_NUM_ID_`[chr.idx]
    <span style="color: #408080; font-style: italic"># calculate LD</span>
    corr0 <span style="color: #666666">&lt;-</span> <span style="color: #0000FF">snp_cor</span>(
        genotype,
        ind.col <span style="color: #666666">=</span> ind.chr,
        ncores <span style="color: #666666">=</span> NCORES,
        infos.pos <span style="color: #666666">=</span> POS2[ind.chr],
        size <span style="color: #666666">=</span> <span style="color: #666666">3</span> <span style="color: #666666">/</span> <span style="color: #666666">1000</span>
    )
    corr <span style="color: #666666">&lt;-</span> bigsparser<span style="color: #666666">::</span><span style="color: #0000FF">as_SFBM</span>(<span style="color: #0000FF">as</span>(corr0, <span style="color: #BA2121">&quot;dgCMatrix&quot;</span>))
    <span style="color: #408080; font-style: italic"># Perform LDSC analysis to get h2 estimate</span>
    ldsc <span style="color: #666666">&lt;-</span> <span style="color: #0000FF">snp_ldsc2</span>(corr0, df_beta)
    h2_est <span style="color: #666666">&lt;-</span> ldsc[[<span style="color: #BA2121">&quot;h2&quot;</span>]]
    <span style="color: #408080; font-style: italic"># Get adjusted beta from the auto model</span>
    multi_auto <span style="color: #666666">&lt;-</span> <span style="color: #0000FF">snp_ldpred2_auto</span>(
        corr,
        df_beta,
        h2_init <span style="color: #666666">=</span> h2_est,
        vec_p_init <span style="color: #666666">=</span> <span style="color: #0000FF">seq_log</span>(<span style="color: #666666">1e-4</span>, <span style="color: #666666">0.9</span>, length.out <span style="color: #666666">=</span> NCORES),
        ncores <span style="color: #666666">=</span> NCORES
    )
    beta_auto <span style="color: #666666">&lt;-</span> <span style="color: #0000FF">sapply</span>(multi_auto, <span style="color: #0000FF">function</span>(auto)
        auto<span style="color: #666666">$</span>beta_est)
    pred_auto <span style="color: #666666">&lt;-</span>
        <span style="color: #0000FF">big_prodMat</span>(genotype,
                    beta_auto,
                    ind.row <span style="color: #666666">=</span> ind.test,
                    ind.col <span style="color: #666666">=</span> ind.chr)
    <span style="color: #408080; font-style: italic"># scale the PRS generated from AUTO</span>
    pred_scaled <span style="color: #666666">&lt;-</span> <span style="color: #0000FF">apply</span>(pred_auto, <span style="color: #666666">2</span>, sd)
    final_beta_auto <span style="color: #666666">&lt;-</span>
        <span style="color: #0000FF">rowMeans</span>(beta_auto[,
                        <span style="color: #0000FF">abs</span>(pred_scaled <span style="color: #666666">-</span>
                                <span style="color: #0000FF">median</span>(pred_scaled)) <span style="color: #666666">&lt;</span>
                            <span style="color: #666666">3</span> <span style="color: #666666">*</span> <span style="color: #0000FF">mad</span>(pred_scaled)])
    pred_auto <span style="color: #666666">&lt;-</span>
        <span style="color: #0000FF">big_prodVec</span>(genotype,
                    final_beta_auto,
                    ind.row <span style="color: #666666">=</span> ind.test,
                    ind.col <span style="color: #666666">=</span> ind.chr)
    <span style="color: #408080; font-style: italic"># add up the calculated auto PRS</span>
    y[, auto <span style="color: #666666">:=</span> auto <span style="color: #666666">+</span> pred_auto]
}
<span style="color: #0000FF">print</span>(<span style="color: #BA2121">&quot;Completed&quot;</span>)
</code></pre></div>

</div>
<input id="__tabbed_5_4" name="__tabbed_5" type="radio" /><label for="__tabbed_5_4">all model</label><div class="tabbed-content">
<div class="highlight" style="background: #f8f8f8"><pre style="line-height: 125%"><span></span><code><span style="color: #408080; font-style: italic"># Get maximum amount of cores</span>
NCORES <span style="color: #666666">&lt;-</span> <span style="color: #0000FF">nb_cores</span>()
<span style="color: #408080; font-style: italic"># Prepare data for grid model</span>
p_seq <span style="color: #666666">&lt;-</span> <span style="color: #0000FF">signif</span>(<span style="color: #0000FF">seq_log</span>(<span style="color: #666666">1e-4</span>, <span style="color: #666666">1</span>, length.out <span style="color: #666666">=</span> <span style="color: #666666">17</span>), <span style="color: #666666">2</span>)
<span style="color: #408080; font-style: italic"># Start doing analysis on each chromosome</span>
y[, Inf.est <span style="color: #666666">:=</span> <span style="color: #666666">0</span>]
y[, auto <span style="color: #666666">:=</span> <span style="color: #666666">0</span>]
y[, grid <span style="color: #666666">:=</span> <span style="color: #666666">0</span>]
<span style="color: #408080; font-style: italic"># add progress bar</span>
pb <span style="color: #666666">=</span> <span style="color: #0000FF">txtProgressBar</span>(min <span style="color: #666666">=</span> <span style="color: #666666">0</span>, max <span style="color: #666666">=</span> <span style="color: #666666">22</span>, initial <span style="color: #666666">=</span> <span style="color: #666666">0</span>)
<span style="color: #0000FF">for </span>(chr in <span style="color: #666666">1:22</span>) {
    <span style="color: #0000FF">setTxtProgressBar</span>(pb, chr)
    <span style="color: #408080; font-style: italic"># Extract current chromosome</span>
    chr.idx <span style="color: #666666">&lt;-</span> <span style="color: #0000FF">which</span>(info_snp<span style="color: #666666">$</span>chr <span style="color: #666666">==</span> chr)
    df_beta <span style="color: #666666">&lt;-</span> info_snp[chr.idx,
                        <span style="color: #0000FF">c</span>(<span style="color: #BA2121">&quot;beta&quot;</span>, <span style="color: #BA2121">&quot;beta_se&quot;</span>, <span style="color: #BA2121">&quot;n_eff&quot;</span>)]
    ind.chr <span style="color: #666666">&lt;-</span> info_snp<span style="color: #666666">$</span>`_NUM_ID_`[chr.idx]
    <span style="color: #408080; font-style: italic"># Calculate LD</span>
    corr0 <span style="color: #666666">&lt;-</span> <span style="color: #0000FF">snp_cor</span>(
        genotype,
        ind.col <span style="color: #666666">=</span> ind.chr,
        ncores <span style="color: #666666">=</span> NCORES,
        infos.pos <span style="color: #666666">=</span> POS2[ind.chr],
        size <span style="color: #666666">=</span> <span style="color: #666666">3</span> <span style="color: #666666">/</span> <span style="color: #666666">1000</span>
    )
    corr <span style="color: #666666">&lt;-</span> bigsparser<span style="color: #666666">::</span><span style="color: #0000FF">as_SFBM</span>(<span style="color: #0000FF">as</span>(corr0, <span style="color: #BA2121">&quot;dgCMatrix&quot;</span>))
    <span style="color: #408080; font-style: italic"># Perform LDSC analysis</span>
    ldsc <span style="color: #666666">&lt;-</span> <span style="color: #0000FF">snp_ldsc2</span>(corr0, df_beta)
    h2_est <span style="color: #666666">&lt;-</span> ldsc[[<span style="color: #BA2121">&quot;h2&quot;</span>]]
    h2_seq <span style="color: #666666">&lt;-</span> <span style="color: #0000FF">round</span>(h2_est <span style="color: #666666">*</span> <span style="color: #0000FF">c</span>(<span style="color: #666666">0.7</span>, <span style="color: #666666">1</span>, <span style="color: #666666">1.4</span>), <span style="color: #666666">4</span>)
    grid.param <span style="color: #666666">&lt;-</span>
        <span style="color: #0000FF">expand.grid</span>(p <span style="color: #666666">=</span> p_seq,
                    h2 <span style="color: #666666">=</span> h2_seq,
                    sparse <span style="color: #666666">=</span> <span style="color: #0000FF">c</span>(<span style="color: #008000; font-weight: bold">FALSE</span>, <span style="color: #008000; font-weight: bold">TRUE</span>))
    <span style="color: #408080; font-style: italic"># Get adjusted beta from infinitesimal model</span>
    beta_inf <span style="color: #666666">&lt;-</span> <span style="color: #0000FF">snp_ldpred2_inf</span>(corr, df_beta, h2 <span style="color: #666666">=</span> h2_est)
    <span style="color: #408080; font-style: italic"># Get adjusted beta from grid model</span>
    beta_grid <span style="color: #666666">&lt;-</span>
        <span style="color: #0000FF">snp_ldpred2_grid</span>(corr, df_beta, grid.param, ncores <span style="color: #666666">=</span> NCORES)
    <span style="color: #408080; font-style: italic"># Get adjusted beta from the auto model</span>
    multi_auto <span style="color: #666666">&lt;-</span> <span style="color: #0000FF">snp_ldpred2_auto</span>(
        corr,
        df_beta,
        h2_init <span style="color: #666666">=</span> h2_est,
        vec_p_init <span style="color: #666666">=</span> <span style="color: #0000FF">seq_log</span>(<span style="color: #666666">1e-4</span>, <span style="color: #666666">0.9</span>, length.out <span style="color: #666666">=</span> NCORES),
        ncores <span style="color: #666666">=</span> NCORES
    )
    beta_auto <span style="color: #666666">&lt;-</span> <span style="color: #0000FF">sapply</span>(multi_auto, <span style="color: #0000FF">function</span>(auto)
        auto<span style="color: #666666">$</span>beta_est)
    pred_auto <span style="color: #666666">&lt;-</span>
        <span style="color: #0000FF">big_prodMat</span>(genotype,
                    beta_auto,
                    ind.row <span style="color: #666666">=</span> ind.test,
                    ind.col <span style="color: #666666">=</span> ind.chr)
    <span style="color: #408080; font-style: italic"># scale the PRS generated from AUTO</span>
    pred_scaled <span style="color: #666666">&lt;-</span> <span style="color: #0000FF">apply</span>(pred_auto, <span style="color: #666666">2</span>, sd)
    final_beta_auto <span style="color: #666666">&lt;-</span>
        <span style="color: #0000FF">rowMeans</span>(beta_auto[,
                        <span style="color: #0000FF">abs</span>(pred_scaled <span style="color: #666666">-</span>
                                <span style="color: #0000FF">median</span>(pred_scaled)) <span style="color: #666666">&lt;</span>
                            <span style="color: #666666">3</span> <span style="color: #666666">*</span> <span style="color: #0000FF">mad</span>(pred_scaled)])
    <span style="color: #408080; font-style: italic"># Get the auto PRS</span>
    pred_auto <span style="color: #666666">&lt;-</span>
        <span style="color: #0000FF">big_prodVec</span>(genotype,
                    final_beta_auto,
                    ind.row <span style="color: #666666">=</span> ind.test,
                    ind.col <span style="color: #666666">=</span> ind.chr)
    <span style="color: #408080; font-style: italic"># Get infinitesimal PRS</span>
    pred_inf <span style="color: #666666">&lt;-</span>
        <span style="color: #0000FF">big_prodVec</span>(genotype,
                    beta_inf,
                    ind.row <span style="color: #666666">=</span> ind.test,
                    ind.col <span style="color: #666666">=</span> ind.chr)
    <span style="color: #408080; font-style: italic"># Get the grid PRSs</span>
    pred_grid <span style="color: #666666">&lt;-</span> <span style="color: #0000FF">big_prodMat</span>(genotype, beta_grid, ind.col <span style="color: #666666">=</span> ind.chr)
    <span style="color: #408080; font-style: italic"># add up the calculated PRS</span>
    y[, Inf.est <span style="color: #666666">:=</span> Inf.est <span style="color: #666666">+</span> pred_inf]
    <span style="color: #408080; font-style: italic"># for grid, we want to retain the &quot;best&quot; prs only</span>
    indep <span style="color: #666666">&lt;-</span> <span style="color: #0000FF">copy</span>(y)
    max.grid.r2 <span style="color: #666666">&lt;-</span> <span style="color: #666666">0</span>
    max.grid.id <span style="color: #666666">&lt;-</span> <span style="color: #666666">0</span>
    <span style="color: #0000FF">for </span>(i in <span style="color: #666666">1:</span><span style="color: #0000FF">ncol</span>(pred_grid)) {
        indep[, PRS <span style="color: #666666">:=</span> pred_grid[, i]]
        model <span style="color: #666666">&lt;-</span> <span style="color: #0000FF">lm</span>(reg.formula, data <span style="color: #666666">=</span> indep) <span style="color: #666666">%&gt;%</span>
            summary
        prs.r2 <span style="color: #666666">&lt;-</span> model<span style="color: #666666">$</span>r.squared <span style="color: #666666">-</span> null.r2
        <span style="color: #0000FF">if </span>(max.grid.r2 <span style="color: #666666">&lt;</span> prs.r2) {
            max.grid.r2 <span style="color: #666666">&lt;-</span> prs.r2
            max.grid.id <span style="color: #666666">&lt;-</span> i
        }
    }
    <span style="color: #408080; font-style: italic"># add up the calculated grid PRS</span>
    y[, grid <span style="color: #666666">:=</span> grid <span style="color: #666666">+</span> pred_grid[, max.grid.id]]
    <span style="color: #408080; font-style: italic"># add up the calculated auto PRS</span>
    y[, auto <span style="color: #666666">:=</span> auto <span style="color: #666666">+</span> pred_auto]
}
<span style="color: #0000FF">print</span>(<span style="color: #BA2121">&quot;Completed&quot;</span>)
</code></pre></div>

</div>
</div>
<h3 id="5-get-the-final-performance-of-the-ldpred-models">5. Get the final performance of the LDpred models<a class="headerlink" href="#5-get-the-final-performance-of-the-ldpred-models" title="Permanent link">&para;</a></h3>
<div class="tabbed-set" data-tabs="6:4"><input checked="checked" id="__tabbed_6_1" name="__tabbed_6" type="radio" /><label for="__tabbed_6_1">infinitesimal model</label><div class="tabbed-content">
<div class="highlight" style="background: #f8f8f8"><pre style="line-height: 125%"><span></span><code><span style="color: #408080; font-style: italic"># there are 3 models: inf, auto and grid</span>
inf.model <span style="color: #666666">&lt;-</span> <span style="color: #0000FF">paste</span>(<span style="color: #BA2121">&quot;PC&quot;</span>, <span style="color: #666666">1:6</span>, sep <span style="color: #666666">=</span> <span style="color: #BA2121">&quot;&quot;</span>, collapse <span style="color: #666666">=</span> <span style="color: #BA2121">&quot;+&quot;</span>) <span style="color: #666666">%&gt;%</span>
    <span style="color: #0000FF">paste0</span>(<span style="color: #BA2121">&quot;Height~Inf.est+Sex+&quot;</span>, .) <span style="color: #666666">%&gt;%</span>
    as.formula <span style="color: #666666">%&gt;%</span>
    <span style="color: #0000FF">lm</span>(., data <span style="color: #666666">=</span> y) <span style="color: #666666">%&gt;%</span>
    summary

result <span style="color: #666666">&lt;-</span> <span style="color: #0000FF">data.table</span>(
    infinitesimal <span style="color: #666666">=</span> inf.model<span style="color: #666666">$</span>r.squared <span style="color: #666666">-</span> null.r2,
    null <span style="color: #666666">=</span> null.r2
)
<span style="color: #0000FF">print</span>(result)
</code></pre></div>

</div>
<input id="__tabbed_6_2" name="__tabbed_6" type="radio" /><label for="__tabbed_6_2">grid model</label><div class="tabbed-content">
<div class="highlight" style="background: #f8f8f8"><pre style="line-height: 125%"><span></span><code>grid.model <span style="color: #666666">&lt;-</span> <span style="color: #0000FF">paste</span>(<span style="color: #BA2121">&quot;PC&quot;</span>, <span style="color: #666666">1:6</span>, sep <span style="color: #666666">=</span> <span style="color: #BA2121">&quot;&quot;</span>, collapse <span style="color: #666666">=</span> <span style="color: #BA2121">&quot;+&quot;</span>) <span style="color: #666666">%&gt;%</span>
    <span style="color: #0000FF">paste0</span>(<span style="color: #BA2121">&quot;Height~grid+Sex+&quot;</span>, .) <span style="color: #666666">%&gt;%</span>
    as.formula <span style="color: #666666">%&gt;%</span>
    <span style="color: #0000FF">lm</span>(., data <span style="color: #666666">=</span> y) <span style="color: #666666">%&gt;%</span>
    summary

result <span style="color: #666666">&lt;-</span> <span style="color: #0000FF">data.table</span>(
    grid <span style="color: #666666">=</span> grid.model<span style="color: #666666">$</span>r.squared <span style="color: #666666">-</span> null.r2,
    null <span style="color: #666666">=</span> null.r2
)
<span style="color: #0000FF">print</span>(result)
</code></pre></div>

</div>
<input id="__tabbed_6_3" name="__tabbed_6" type="radio" /><label for="__tabbed_6_3">auto model</label><div class="tabbed-content">
<div class="highlight" style="background: #f8f8f8"><pre style="line-height: 125%"><span></span><code>auto.model <span style="color: #666666">&lt;-</span> <span style="color: #0000FF">paste</span>(<span style="color: #BA2121">&quot;PC&quot;</span>, <span style="color: #666666">1:6</span>, sep <span style="color: #666666">=</span> <span style="color: #BA2121">&quot;&quot;</span>, collapse <span style="color: #666666">=</span> <span style="color: #BA2121">&quot;+&quot;</span>) <span style="color: #666666">%&gt;%</span>
    <span style="color: #0000FF">paste0</span>(<span style="color: #BA2121">&quot;Height~auto+Sex+&quot;</span>, .) <span style="color: #666666">%&gt;%</span>
    as.formula <span style="color: #666666">%&gt;%</span>
    <span style="color: #0000FF">lm</span>(., data <span style="color: #666666">=</span> y) <span style="color: #666666">%&gt;%</span>
    summary

result <span style="color: #666666">&lt;-</span> <span style="color: #0000FF">data.table</span>(
    infinitesimal <span style="color: #666666">=</span> inf.model<span style="color: #666666">$</span>r.squared <span style="color: #666666">-</span> null.r2,
    grid <span style="color: #666666">=</span> grid.model<span style="color: #666666">$</span>r.squared <span style="color: #666666">-</span> null.r2,
    auto <span style="color: #666666">=</span> auto.model<span style="color: #666666">$</span>r.squared <span style="color: #666666">-</span> null.r2,
    null <span style="color: #666666">=</span> null.r2
)
<span style="color: #0000FF">print</span>(result)
</code></pre></div>

</div>
<input id="__tabbed_6_4" name="__tabbed_6" type="radio" /><label for="__tabbed_6_4">all model</label><div class="tabbed-content">
<div class="highlight" style="background: #f8f8f8"><pre style="line-height: 125%"><span></span><code><span style="color: #408080; font-style: italic"># there are 3 models: inf, auto and grid</span>
inf.model <span style="color: #666666">&lt;-</span> <span style="color: #0000FF">paste</span>(<span style="color: #BA2121">&quot;PC&quot;</span>, <span style="color: #666666">1:6</span>, sep <span style="color: #666666">=</span> <span style="color: #BA2121">&quot;&quot;</span>, collapse <span style="color: #666666">=</span> <span style="color: #BA2121">&quot;+&quot;</span>) <span style="color: #666666">%&gt;%</span>
    <span style="color: #0000FF">paste0</span>(<span style="color: #BA2121">&quot;Height~Inf.est+Sex+&quot;</span>, .) <span style="color: #666666">%&gt;%</span>
    as.formula <span style="color: #666666">%&gt;%</span>
    <span style="color: #0000FF">lm</span>(., data <span style="color: #666666">=</span> y) <span style="color: #666666">%&gt;%</span>
    summary
auto.model <span style="color: #666666">&lt;-</span> <span style="color: #0000FF">paste</span>(<span style="color: #BA2121">&quot;PC&quot;</span>, <span style="color: #666666">1:6</span>, sep <span style="color: #666666">=</span> <span style="color: #BA2121">&quot;&quot;</span>, collapse <span style="color: #666666">=</span> <span style="color: #BA2121">&quot;+&quot;</span>) <span style="color: #666666">%&gt;%</span>
    <span style="color: #0000FF">paste0</span>(<span style="color: #BA2121">&quot;Height~auto+Sex+&quot;</span>, .) <span style="color: #666666">%&gt;%</span>
    as.formula <span style="color: #666666">%&gt;%</span>
    <span style="color: #0000FF">lm</span>(., data <span style="color: #666666">=</span> y) <span style="color: #666666">%&gt;%</span>
    summary
grid.model <span style="color: #666666">&lt;-</span> <span style="color: #0000FF">paste</span>(<span style="color: #BA2121">&quot;PC&quot;</span>, <span style="color: #666666">1:6</span>, sep <span style="color: #666666">=</span> <span style="color: #BA2121">&quot;&quot;</span>, collapse <span style="color: #666666">=</span> <span style="color: #BA2121">&quot;+&quot;</span>) <span style="color: #666666">%&gt;%</span>
    <span style="color: #0000FF">paste0</span>(<span style="color: #BA2121">&quot;Height~grid+Sex+&quot;</span>, .) <span style="color: #666666">%&gt;%</span>
    as.formula <span style="color: #666666">%&gt;%</span>
    <span style="color: #0000FF">lm</span>(., data <span style="color: #666666">=</span> y) <span style="color: #666666">%&gt;%</span>
    summary

result <span style="color: #666666">&lt;-</span> <span style="color: #0000FF">data.table</span>(
    infinitesimal <span style="color: #666666">=</span> inf.model<span style="color: #666666">$</span>r.squared <span style="color: #666666">-</span> null.r2,
    grid <span style="color: #666666">=</span> grid.model<span style="color: #666666">$</span>r.squared <span style="color: #666666">-</span> null.r2,
    auto <span style="color: #666666">=</span> auto.model<span style="color: #666666">$</span>r.squared <span style="color: #666666">-</span> null.r2,
    null <span style="color: #666666">=</span> null.r2
)
<span style="color: #0000FF">print</span>(result)
</code></pre></div>

</div>
</div>
<details class="note"><summary>How much phenotypic variation does the PRS from each model explain?</summary><p>Infinitesimal = 0.1717342</p>
<p>Grid Model = 0.002945278</p>
<p>Auto Model = 0.2532782</p>
</details>
              
            </div>
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="../lassosum/" class="btn btn-neutral float-right" title="lassosum">Next <span class="icon icon-circle-arrow-right"></span></a>
      
      
        <a href="../prsice/" class="btn btn-neutral" title="PRSice-2"><span class="icon icon-circle-arrow-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <!-- Copyright etc -->
    
  </div>

  Built with <a href="https://www.mkdocs.org/">MkDocs</a> using a <a href="https://github.com/snide/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>.
</footer>
      
        </div>
      </div>

    </section>

  </div>

  <div class="rst-versions" role="note" aria-label="versions">
    <span class="rst-current-version" data-toggle="rst-current-version">
      
          <a href="https://github.com/choishingwan/PRS-Tutorial/" class="fa fa-github" style="float: left; color: #fcfcfc"> GitHub</a>
      
      
        <span><a href="../prsice/" style="color: #fcfcfc;">&laquo; Previous</a></span>
      
      
        <span style="margin-left: 15px"><a href="../lassosum/" style="color: #fcfcfc">Next &raquo;</a></span>
      
    </span>
</div>
    <script>var base_url = '..';</script>
    <script src="../js/theme.js" defer></script>
      <script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.0/MathJax.js?config=TeX-MML-AM_CHTML" defer></script>
      <script src="../javascripts/details.js" defer></script>
      <script src="../search/main.js" defer></script>
    <script defer>
        window.onload = function () {
            SphinxRtdTheme.Navigation.enable(true);
        };
    </script>

</body>
</html>
