<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <meta name="author" content="Shing Wan CHOI">
  <link rel="shortcut icon" href="../img/favicon.ico">
  <title>LDpred - Basic Tutorial for Polygenic Risk Score Analyses</title>
  <link href='https://fonts.googleapis.com/css?family=Lato:400,700|Roboto+Slab:400,700|Inconsolata:400,700' rel='stylesheet' type='text/css'>

  <link rel="stylesheet" href="../css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../css/theme_extra.css" type="text/css" />
  <link rel="stylesheet" href="//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/styles/github.min.css">
  <link href="../css/extra.css" rel="stylesheet">
  
  <script>
    // Current page data
    var mkdocs_page_name = "LDpred";
    var mkdocs_page_input_path = "ldpred.md";
    var mkdocs_page_url = null;
  </script>
  
  <script src="../js/jquery-2.1.1.min.js" defer></script>
  <script src="../js/modernizr-2.8.3.min.js" defer></script>
  <script src="//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/highlight.min.js"></script>
  <script>hljs.initHighlightingOnLoad();</script> 
  
  <script>
      (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
      (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
      m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
      })(window,document,'script','https://www.google-analytics.com/analytics.js','ga');

      ga('create', 'UA-145068952-1', 'auto');
      ga('send', 'pageview');
  </script>
  
</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side stickynav">
      <div class="wy-side-nav-search">
        <a href=".." class="icon icon-home"> Basic Tutorial for Polygenic Risk Score Analyses</a>
        <div role="search">
  <form id ="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" title="Type search term here" />
  </form>
</div>
      </div>

      <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
	<ul class="current">
	  
          
            <li class="toctree-l1">
		
    <a class="" href="..">Overview</a>
	    </li>
          
            <li class="toctree-l1">
		
    <a class="" href="../base/">1. QC of Base Data</a>
	    </li>
          
            <li class="toctree-l1">
		
    <a class="" href="../target/">2. QC of Target Data</a>
	    </li>
          
            <li class="toctree-l1">
		
    <span class="caption-text">3. Calculating and analysing PRS</span>
    <ul class="subnav">
                <li class="">
                    
    <a class="" href="../plink/">PLINK</a>
                </li>
                <li class="">
                    
    <a class="" href="../prsice/">PRSice-2</a>
                </li>
                <li class=" current">
                    
    <a class="current" href="./">LDpred</a>
    <ul class="subnav">
            
    <li class="toctree-l3"><a href="#running-prs-analysis">Running PRS analysis</a></li>
    

    <li class="toctree-l3"><a href="#0-on-some-server-you-might-need-to-first-use-the-following-code-in-order-to-run-ldpred-with-multi-thread">0. On some server, you might need to first use the following code in order to run LDpred with multi-thread</a></li>
    

    <li class="toctree-l3"><a href="#1-read-in-the-phenotype-and-covariate-files">1. Read in the phenotype and covariate files</a></li>
    

    <li class="toctree-l3"><a href="#2-load-in-the-genotype-and-summary-statistic-file-using-bigsnpr">2. Load in the genotype and summary statistic file using bigsnpr</a></li>
    

    <li class="toctree-l3"><a href="#3-reformat-data-in-preparation-of-ldpred2">3. Reformat data in preparation of LDpred2</a></li>
    

    <li class="toctree-l3"><a href="#4-start-running-ldpred-2">4. Start running LDpred 2</a></li>
    

    <li class="toctree-l3"><a href="#5-get-the-final-performance-of-the-ldpred-models">5. Get the final performance of the LDpred models</a></li>
    

    </ul>
                </li>
                <li class="">
                    
    <a class="" href="../lassosum/">lassosum</a>
                </li>
    </ul>
	    </li>
          
            <li class="toctree-l1">
		
    <a class="" href="../plink_visual/">4. Visualizing PRS Results</a>
	    </li>
          
        </ul>
      </div>
      &nbsp;
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" role="navigation" aria-label="top navigation">
        <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
        <a href="..">Basic Tutorial for Polygenic Risk Score Analyses</a>
      </nav>

      
      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href="..">Docs</a> &raquo;</li>
    
      
        
          <li>3. Calculating and analysing PRS &raquo;</li>
        
      
    
    <li>LDpred</li>
    <li class="wy-breadcrumbs-aside">
      
        <a href="https://github.com/choishingwan/PRS-Tutorial/edit/master/docs/ldpred.md"
          class="icon icon-github"> Edit on GitHub</a>
      
    </li>
  </ul>
  <hr/>
</div>
          <div role="main">
            <div class="section">
              
                <p>Here we use another PRS program, <a href="https://privefl.github.io/bigsnpr/articles/LDpred2.html">LDpred-2</a>, that uses a Bayesian approach to polygenic risk scoring.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The script used here is based on LDpred 2 implemented under bigsnpr version 1.4.7</p>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>For more details, please refer to <a href="https://privefl.github.io/bigsnpr/articles/LDpred2.html">LDpred 2's homepage</a></p>
</div>
<p>You can install <code>LDpred</code> and its dependencies in <code>R</code> with the following command:</p>
<div class="codehilite"><pre><span></span>install.packages<span class="p">(</span><span class="s">&quot;remotes&quot;</span><span class="p">)</span>
<span class="kn">library</span><span class="p">(</span>remotes<span class="p">)</span>
remotes<span class="o">::</span>install_github<span class="p">(</span><span class="s">&quot;https://github.com/privefl/bigsnpr.git&quot;</span><span class="p">)</span>
</pre></div>

<div class="admonition note">
<p class="admonition-title">Note</p>
<p>For mac users, you might need to follow the guide <a href="https://thecoatlessprofessor.com/programming/cpp/openmp-in-r-on-os-x/">here</a> to be able to install LDpred2</p>
</div>
<p>We assume that you have the following files (or you can download it from <a href="https://drive.google.com/file/d/1x_G0Gxk9jFMY-PMqwtg6-vdEyUPp5p5u/view?usp=sharing">here</a>):</p>
<table>
<thead>
<tr>
<th align="center">File Name</th>
<th align="center">Description</th>
</tr>
</thead>
<tbody>
<tr>
<td align="center"><strong>Height.QC.gz</strong></td>
<td align="center">The post-QCed summary statistic</td>
</tr>
<tr>
<td align="center"><strong>EUR.QC.bed</strong></td>
<td align="center">The genotype file after performing some basic filtering</td>
</tr>
<tr>
<td align="center"><strong>EUR.QC.bim</strong></td>
<td align="center">This file contains the SNPs that passed the basic filtering</td>
</tr>
<tr>
<td align="center"><strong>EUR.QC.fam</strong></td>
<td align="center">This file contains the samples that passed the basic filtering</td>
</tr>
<tr>
<td align="center"><strong>EUR.height</strong></td>
<td align="center">This file contains the phenotype of the samples</td>
</tr>
<tr>
<td align="center"><strong>EUR.cov</strong></td>
<td align="center">This file contains the covariates of the samples</td>
</tr>
<tr>
<td align="center"><strong>EUR.eigenvec</strong></td>
<td align="center">This file contains the PCs of the samples</td>
</tr>
</tbody>
</table>
<h1 id="running-prs-analysis">Running PRS analysis<a class="headerlink" href="#running-prs-analysis" title="Permanent link">&para;</a></h1>
<h1 id="0-on-some-server-you-might-need-to-first-use-the-following-code-in-order-to-run-ldpred-with-multi-thread">0. On some server, you might need to first use the following code in order to run LDpred with multi-thread<a class="headerlink" href="#0-on-some-server-you-might-need-to-first-use-the-following-code-in-order-to-run-ldpred-with-multi-thread" title="Permanent link">&para;</a></h1>
<div class="superfences-tabs">
<input name="__tabs_1" type="radio" id="__tab_1_0" checked="checked" />
<label for="__tab_1_0">prepare workspace</label>
<div class="superfences-content"><div class="codehilite"><pre><span></span><span class="kp">options</span><span class="p">(</span>bigstatsr.check.parallel.blas <span class="o">=</span> <span class="kc">FALSE</span><span class="p">)</span>
<span class="kp">options</span><span class="p">(</span>default.nproc.blas <span class="o">=</span> <span class="kc">NULL</span><span class="p">)</span>
</pre></div></div>
</div>
<h1 id="1-read-in-the-phenotype-and-covariate-files">1. Read in the phenotype and covariate files<a class="headerlink" href="#1-read-in-the-phenotype-and-covariate-files" title="Permanent link">&para;</a></h1>
<div class="superfences-tabs">
<input name="__tabs_2" type="radio" id="__tab_2_0" checked="checked" />
<label for="__tab_2_0">read in phenotype and covariates</label>
<div class="superfences-content"><div class="codehilite"><pre><span></span><span class="kn">library</span><span class="p">(</span>data.table<span class="p">)</span>
<span class="kn">library</span><span class="p">(</span>magrittr<span class="p">)</span>
phenotype <span class="o">&lt;-</span> fread<span class="p">(</span><span class="s">&quot;EUR.height&quot;</span><span class="p">)</span>
covariate <span class="o">&lt;-</span> fread<span class="p">(</span><span class="s">&quot;EUR.cov&quot;</span><span class="p">)</span>
pcs <span class="o">&lt;-</span> fread<span class="p">(</span><span class="s">&quot;EUR.eigenvec&quot;</span><span class="p">)</span>
<span class="c1"># rename columns</span>
<span class="kp">colnames</span><span class="p">(</span>pcs<span class="p">)</span> <span class="o">&lt;-</span> <span class="kt">c</span><span class="p">(</span><span class="s">&quot;FID&quot;</span><span class="p">,</span><span class="s">&quot;IID&quot;</span><span class="p">,</span> <span class="kp">paste0</span><span class="p">(</span><span class="s">&quot;PC&quot;</span><span class="p">,</span><span class="m">1</span><span class="o">:</span><span class="m">6</span><span class="p">))</span>
<span class="c1"># generate required table</span>
pheno <span class="o">&lt;-</span> <span class="kp">merge</span><span class="p">(</span>phenotype<span class="p">,</span> covariate<span class="p">)</span> <span class="o">%&gt;%</span>
    <span class="kp">merge</span><span class="p">(</span><span class="m">.</span><span class="p">,</span> pcs<span class="p">)</span>
</pre></div></div>
</div>
<h1 id="2-load-in-the-genotype-and-summary-statistic-file-using-bigsnpr">2. Load in the genotype and summary statistic file using bigsnpr<a class="headerlink" href="#2-load-in-the-genotype-and-summary-statistic-file-using-bigsnpr" title="Permanent link">&para;</a></h1>
<div class="superfences-tabs">
<input name="__tabs_3" type="radio" id="__tab_3_0" checked="checked" />
<label for="__tab_3_0">Load genotype with bigsnpr</label>
<div class="superfences-content"><div class="codehilite"><pre><span></span><span class="c1"># load the library</span>
<span class="kn">library</span><span class="p">(</span>bigsnpr<span class="p">)</span>
<span class="c1"># now preprocess the bed file. Only need to do once for each data set</span>
snp_readBed<span class="p">(</span><span class="s">&quot;EUR.QC.bed&quot;</span><span class="p">)</span>
<span class="c1"># now attach the genotype object</span>
obj.bigSNP <span class="o">&lt;-</span> snp_attach<span class="p">(</span><span class="s">&quot;EUR.QC.rds&quot;</span><span class="p">)</span>
<span class="c1"># Next, read in the summary statistic file</span>
sumstats <span class="o">&lt;-</span> bigreadr<span class="o">::</span>fread2<span class="p">(</span><span class="s">&quot;Height.QC.gz&quot;</span><span class="p">)</span>
</pre></div></div>
</div>
<h1 id="3-reformat-data-in-preparation-of-ldpred2">3. Reformat data in preparation of LDpred2<a class="headerlink" href="#3-reformat-data-in-preparation-of-ldpred2" title="Permanent link">&para;</a></h1>
<div class="superfences-tabs">
<input name="__tabs_4" type="radio" id="__tab_4_0" checked="checked" />
<label for="__tab_4_0">data reformating</label>
<div class="superfences-content"><div class="codehilite"><pre><span></span><span class="c1"># LDpred 2 require the header to follow the exact naming</span>
<span class="kp">names</span><span class="p">(</span>sumstats<span class="p">)</span> <span class="o">&lt;-</span>
    <span class="kt">c</span><span class="p">(</span><span class="s">&quot;chr&quot;</span><span class="p">,</span>
      <span class="s">&quot;pos&quot;</span><span class="p">,</span>
      <span class="s">&quot;rsid&quot;</span><span class="p">,</span>
      <span class="s">&quot;a1&quot;</span><span class="p">,</span>
      <span class="s">&quot;a0&quot;</span><span class="p">,</span>
      <span class="s">&quot;n_eff&quot;</span><span class="p">,</span>
      <span class="s">&quot;beta_se&quot;</span><span class="p">,</span>
      <span class="s">&quot;p&quot;</span><span class="p">,</span>
      <span class="s">&quot;OR&quot;</span><span class="p">,</span>
      <span class="s">&quot;INFO&quot;</span><span class="p">,</span>
      <span class="s">&quot;MAF&quot;</span><span class="p">)</span>
<span class="c1"># Transform the OR into log(OR)</span>
sumstats<span class="o">$</span>beta <span class="o">&lt;-</span> <span class="kp">log</span><span class="p">(</span>sumstats<span class="o">$</span>OR<span class="p">)</span>
<span class="c1"># extract the SNP information from the genotype</span>
map <span class="o">&lt;-</span> obj.bigSNP<span class="o">$</span>map<span class="p">[</span><span class="m">-2</span><span class="p">]</span>
<span class="kp">names</span><span class="p">(</span>map<span class="p">)</span> <span class="o">&lt;-</span> <span class="kt">c</span><span class="p">(</span><span class="s">&quot;chr&quot;</span><span class="p">,</span> <span class="s">&quot;rsid&quot;</span><span class="p">,</span> <span class="s">&quot;pos&quot;</span><span class="p">,</span> <span class="s">&quot;a1&quot;</span><span class="p">,</span> <span class="s">&quot;a0&quot;</span><span class="p">)</span>
<span class="c1"># perform SNP matching</span>
info_snp <span class="o">&lt;-</span> snp_match<span class="p">(</span>sumstats<span class="p">,</span> map<span class="p">)</span>
<span class="c1"># Rename the data structures</span>
CHR <span class="o">&lt;-</span> map<span class="o">$</span>chr
POS <span class="o">&lt;-</span> map<span class="o">$</span>pos
<span class="c1"># get the CM information from 1000 Genome</span>
<span class="c1"># will download the 1000G file to the current directory (&quot;.&quot;)</span>
POS2 <span class="o">&lt;-</span> snp_asGeneticPos<span class="p">(</span>CHR<span class="p">,</span> POS<span class="p">,</span> dir <span class="o">=</span> <span class="s">&quot;.&quot;</span><span class="p">)</span>
<span class="c1"># Reformat the phenotype information to match the fam order</span>
fam.order <span class="o">&lt;-</span> as.data.table<span class="p">(</span>obj.bigSNP<span class="o">$</span>fam<span class="p">)</span>
setnames<span class="p">(</span>fam.order<span class="p">,</span>
         <span class="kt">c</span><span class="p">(</span><span class="s">&quot;family.ID&quot;</span><span class="p">,</span> <span class="s">&quot;sample.ID&quot;</span><span class="p">),</span>
         <span class="kt">c</span><span class="p">(</span><span class="s">&quot;FID&quot;</span><span class="p">,</span> <span class="s">&quot;IID&quot;</span><span class="p">))</span>
<span class="c1"># This will ensure our phenotype vector y is of the same order as the</span>
<span class="c1"># family object generated by bigsnpr</span>
y <span class="o">&lt;-</span> pheno<span class="p">[</span>fam.order<span class="p">,</span> on <span class="o">=</span> <span class="kt">c</span><span class="p">(</span><span class="s">&quot;FID&quot;</span><span class="p">,</span> <span class="s">&quot;IID&quot;</span><span class="p">)]</span>
genotype <span class="o">&lt;-</span> obj.bigSNP<span class="o">$</span>genotypes
<span class="c1"># Use all sample for test</span>
ind.test <span class="o">&lt;-</span> <span class="m">1</span><span class="o">:</span><span class="kp">nrow</span><span class="p">(</span>genotype<span class="p">)</span>
<span class="c1"># Calculate the null R2</span>
<span class="c1"># use glm for binary trait </span>
<span class="c1"># (will also need the fmsb package to calculate the pseudo R2)</span>
null.model <span class="o">&lt;-</span> <span class="kp">paste</span><span class="p">(</span><span class="s">&quot;PC&quot;</span><span class="p">,</span> <span class="m">1</span><span class="o">:</span><span class="m">6</span><span class="p">,</span> sep <span class="o">=</span> <span class="s">&quot;&quot;</span><span class="p">,</span> collapse <span class="o">=</span> <span class="s">&quot;+&quot;</span><span class="p">)</span> <span class="o">%&gt;%</span>
    <span class="kp">paste0</span><span class="p">(</span><span class="s">&quot;Height~Sex+&quot;</span><span class="p">,</span> <span class="m">.</span><span class="p">)</span> <span class="o">%&gt;%</span>
    as.formula <span class="o">%&gt;%</span>
    lm<span class="p">(</span><span class="m">.</span><span class="p">,</span> data <span class="o">=</span> y<span class="p">)</span> <span class="o">%&gt;%</span>
    <span class="kp">summary</span>
null.r2 <span class="o">&lt;-</span> null.model<span class="o">$</span>r.squared
reg.formula <span class="o">&lt;-</span> <span class="kp">paste</span><span class="p">(</span><span class="s">&quot;PC&quot;</span><span class="p">,</span> <span class="m">1</span><span class="o">:</span><span class="m">6</span><span class="p">,</span> sep <span class="o">=</span> <span class="s">&quot;&quot;</span><span class="p">,</span> collapse <span class="o">=</span> <span class="s">&quot;+&quot;</span><span class="p">)</span> <span class="o">%&gt;%</span>
    <span class="kp">paste0</span><span class="p">(</span><span class="s">&quot;Height~PRS+Sex+&quot;</span><span class="p">,</span> <span class="m">.</span><span class="p">)</span> <span class="o">%&gt;%</span>
    as.formula
</pre></div></div>
</div>
<h1 id="4-start-running-ldpred-2">4. Start running LDpred 2<a class="headerlink" href="#4-start-running-ldpred-2" title="Permanent link">&para;</a></h1>
<div class="superfences-tabs">
<input name="__tabs_5" type="radio" id="__tab_5_0" checked="checked" />
<label for="__tab_5_0">infinitesimal model</label>
<div class="superfences-content"><div class="codehilite"><pre><span></span><span class="c1"># Get maximum amount of cores</span>
NCORES <span class="o">&lt;-</span> nb_cores<span class="p">()</span>
<span class="c1"># Start doing analysis on each chromosome</span>
y<span class="p">[,</span> Inf.est <span class="o">:=</span> <span class="m">0</span><span class="p">]</span>
<span class="c1"># add progress bar</span>
pb <span class="o">=</span> txtProgressBar<span class="p">(</span>min <span class="o">=</span> <span class="m">0</span><span class="p">,</span> max <span class="o">=</span> <span class="m">22</span><span class="p">,</span> initial <span class="o">=</span> <span class="m">0</span><span class="p">)</span>
<span class="kr">for</span> <span class="p">(</span>chr <span class="kr">in</span> <span class="m">1</span><span class="o">:</span><span class="m">22</span><span class="p">)</span> <span class="p">{</span>
    setTxtProgressBar<span class="p">(</span>pb<span class="p">,</span> chr<span class="p">)</span>
    <span class="c1"># extract current chromosome</span>
    chr.idx <span class="o">&lt;-</span> <span class="kp">which</span><span class="p">(</span>info_snp<span class="o">$</span>chr <span class="o">==</span> chr<span class="p">)</span>
    df_beta <span class="o">&lt;-</span> info_snp<span class="p">[</span>chr.idx<span class="p">,</span>
                        <span class="kt">c</span><span class="p">(</span><span class="s">&quot;beta&quot;</span><span class="p">,</span> <span class="s">&quot;beta_se&quot;</span><span class="p">,</span> <span class="s">&quot;n_eff&quot;</span><span class="p">)]</span>
    ind.chr <span class="o">&lt;-</span> info_snp<span class="o">$</span><span class="sb">`_NUM_ID_`</span><span class="p">[</span>chr.idx<span class="p">]</span>
    <span class="c1"># calculate LD</span>
    corr0 <span class="o">&lt;-</span> snp_cor<span class="p">(</span>
        genotype<span class="p">,</span>
        ind.col <span class="o">=</span> ind.chr<span class="p">,</span>
        ncores <span class="o">=</span> NCORES<span class="p">,</span>
        infos.pos <span class="o">=</span> POS2<span class="p">[</span>ind.chr<span class="p">],</span>
        size <span class="o">=</span> <span class="m">3</span> <span class="o">/</span> <span class="m">1000</span>
    <span class="p">)</span>
    corr <span class="o">&lt;-</span> bigsparser<span class="o">::</span>as_SFBM<span class="p">(</span>as<span class="p">(</span>corr0<span class="p">,</span> <span class="s">&quot;dgCMatrix&quot;</span><span class="p">))</span>
    <span class="c1"># Perform LDSC analysis to get h2 estimate</span>
    ldsc <span class="o">&lt;-</span> snp_ldsc2<span class="p">(</span>corr0<span class="p">,</span> df_beta<span class="p">)</span>
    h2_est <span class="o">&lt;-</span> ldsc<span class="p">[[</span><span class="s">&quot;h2&quot;</span><span class="p">]]</span>
    <span class="c1"># Get adjusted beta from infinitesimal model</span>
    beta_inf <span class="o">&lt;-</span> snp_ldpred2_inf<span class="p">(</span>corr<span class="p">,</span> df_beta<span class="p">,</span> h2 <span class="o">=</span> h2_est<span class="p">)</span>
    <span class="c1"># Get infinitesimal PRS</span>
    pred_inf <span class="o">&lt;-</span> big_prodVec<span class="p">(</span>genotype<span class="p">,</span>
                            beta_inf<span class="p">,</span>
                            ind.row <span class="o">=</span> ind.test<span class="p">,</span>
                            ind.col <span class="o">=</span> ind.chr<span class="p">)</span>
    <span class="c1"># add up the calculated PRS</span>
    y<span class="p">[,</span> Inf.est <span class="o">:=</span> Inf.est <span class="o">+</span> pred_inf<span class="p">]</span>
<span class="p">}</span>
<span class="kp">print</span><span class="p">(</span><span class="s">&quot;Completed&quot;</span><span class="p">)</span>
</pre></div></div>
<input name="__tabs_5" type="radio" id="__tab_5_1" />
<label for="__tab_5_1">grid model</label>
<div class="superfences-content"><div class="codehilite"><pre><span></span><span class="c1"># Get maximum amount of cores</span>
NCORES <span class="o">&lt;-</span> nb_cores<span class="p">()</span>
<span class="c1"># Prepare data for grid model</span>
p_seq <span class="o">&lt;-</span> <span class="kp">signif</span><span class="p">(</span>seq_log<span class="p">(</span><span class="m">1e-4</span><span class="p">,</span> <span class="m">1</span><span class="p">,</span> length.out <span class="o">=</span> <span class="m">17</span><span class="p">),</span> <span class="m">2</span><span class="p">)</span>
<span class="c1"># Start doing analysis on each chromosome</span>
y<span class="p">[,</span> grid <span class="o">:=</span> <span class="m">0</span><span class="p">]</span>
<span class="c1"># add progress bar</span>
pb <span class="o">=</span> txtProgressBar<span class="p">(</span>min <span class="o">=</span> <span class="m">0</span><span class="p">,</span> max <span class="o">=</span> <span class="m">22</span><span class="p">,</span> initial <span class="o">=</span> <span class="m">0</span><span class="p">)</span>
<span class="kr">for</span> <span class="p">(</span>chr <span class="kr">in</span> <span class="m">1</span><span class="o">:</span><span class="m">22</span><span class="p">)</span> <span class="p">{</span>
    setTxtProgressBar<span class="p">(</span>pb<span class="p">,</span> chr<span class="p">)</span>
    <span class="c1"># extract current chromosome</span>
    chr.idx <span class="o">&lt;-</span> <span class="kp">which</span><span class="p">(</span>info_snp<span class="o">$</span>chr <span class="o">==</span> chr<span class="p">)</span>
    df_beta <span class="o">&lt;-</span> info_snp<span class="p">[</span>chr.idx<span class="p">,</span>
                        <span class="kt">c</span><span class="p">(</span><span class="s">&quot;beta&quot;</span><span class="p">,</span> <span class="s">&quot;beta_se&quot;</span><span class="p">,</span> <span class="s">&quot;n_eff&quot;</span><span class="p">)]</span>
    ind.chr <span class="o">&lt;-</span> info_snp<span class="o">$</span><span class="sb">`_NUM_ID_`</span><span class="p">[</span>chr.idx<span class="p">]</span>
    <span class="c1"># calculate LD</span>
    corr0 <span class="o">&lt;-</span> snp_cor<span class="p">(</span>
        genotype<span class="p">,</span>
        ind.col <span class="o">=</span> ind.chr<span class="p">,</span>
        ncores <span class="o">=</span> NCORES<span class="p">,</span>
        infos.pos <span class="o">=</span> POS2<span class="p">[</span>ind.chr<span class="p">],</span>
        size <span class="o">=</span> <span class="m">3</span> <span class="o">/</span> <span class="m">1000</span>
    <span class="p">)</span>
    corr <span class="o">&lt;-</span> bigsparser<span class="o">::</span>as_SFBM<span class="p">(</span>as<span class="p">(</span>corr0<span class="p">,</span> <span class="s">&quot;dgCMatrix&quot;</span><span class="p">))</span>
    <span class="c1"># Perform LDSC analysis to get h2 estimate</span>
    ldsc <span class="o">&lt;-</span> snp_ldsc2<span class="p">(</span>corr0<span class="p">,</span> df_beta<span class="p">)</span>
    h2_est <span class="o">&lt;-</span> ldsc<span class="p">[[</span><span class="s">&quot;h2&quot;</span><span class="p">]]</span>
    h2_seq <span class="o">&lt;-</span> <span class="kp">round</span><span class="p">(</span>h2_est <span class="o">*</span> <span class="kt">c</span><span class="p">(</span><span class="m">0.7</span><span class="p">,</span> <span class="m">1</span><span class="p">,</span> <span class="m">1.4</span><span class="p">),</span> <span class="m">4</span><span class="p">)</span>
    grid.param <span class="o">&lt;-</span>
        <span class="kp">expand.grid</span><span class="p">(</span>p <span class="o">=</span> p_seq<span class="p">,</span>
                    h2 <span class="o">=</span> h2_seq<span class="p">,</span>
                    sparse <span class="o">=</span> <span class="kt">c</span><span class="p">(</span><span class="kc">FALSE</span><span class="p">,</span> <span class="kc">TRUE</span><span class="p">))</span>
    <span class="c1"># Get adjusted beta from grid model</span>
    beta_grid <span class="o">&lt;-</span>
        snp_ldpred2_grid<span class="p">(</span>corr<span class="p">,</span> df_beta<span class="p">,</span> grid.param<span class="p">,</span> ncores <span class="o">=</span> NCORES<span class="p">)</span>
    <span class="c1"># Get the grid PRSs</span>
    pred_grid <span class="o">&lt;-</span> big_prodMat<span class="p">(</span>genotype<span class="p">,</span> beta_grid<span class="p">,</span> ind.col <span class="o">=</span> ind.chr<span class="p">)</span>
    <span class="c1"># Find the current &quot;best&quot; PRS</span>
    indep <span class="o">&lt;-</span> copy<span class="p">(</span>y<span class="p">)</span>
    max.grid.r2 <span class="o">&lt;-</span> <span class="m">0</span>
    max.grid.id <span class="o">&lt;-</span> <span class="m">0</span>
    <span class="kr">for</span> <span class="p">(</span>i <span class="kr">in</span> <span class="m">1</span><span class="o">:</span><span class="kp">ncol</span><span class="p">(</span>pred_grid<span class="p">))</span> <span class="p">{</span>
        indep<span class="p">[,</span> PRS <span class="o">:=</span> pred_grid<span class="p">[,</span> i<span class="p">]]</span>
        model <span class="o">&lt;-</span> lm<span class="p">(</span>reg.formula<span class="p">,</span> data <span class="o">=</span> indep<span class="p">)</span> <span class="o">%&gt;%</span>
            <span class="kp">summary</span>
        prs.r2 <span class="o">&lt;-</span> model<span class="o">$</span>r.squared <span class="o">-</span> null.r2
        <span class="kr">if</span> <span class="p">(</span>max.grid.r2 <span class="o">&lt;</span> prs.r2<span class="p">)</span> <span class="p">{</span>
            max.grid.r2 <span class="o">&lt;-</span> prs.r2
            max.grid.id <span class="o">&lt;-</span> i
        <span class="p">}</span>
    <span class="p">}</span>
    <span class="c1"># add up the calculated grid PRS</span>
    y<span class="p">[,</span> grid <span class="o">:=</span> grid <span class="o">+</span> pred_grid<span class="p">[,</span> max.grid.id<span class="p">]]</span>
<span class="p">}</span>
<span class="kp">print</span><span class="p">(</span><span class="s">&quot;Completed&quot;</span><span class="p">)</span>
</pre></div></div>
<input name="__tabs_5" type="radio" id="__tab_5_2" />
<label for="__tab_5_2">auto model</label>
<div class="superfences-content"><div class="codehilite"><pre><span></span><span class="c1"># Get maximum amount of cores</span>
NCORES <span class="o">&lt;-</span> nb_cores<span class="p">()</span>
<span class="c1"># Start doing analysis on each chromosome</span>
y<span class="p">[,</span> auto <span class="o">:=</span> <span class="m">0</span><span class="p">]</span>
<span class="c1"># add progress bar</span>
pb <span class="o">=</span> txtProgressBar<span class="p">(</span>min <span class="o">=</span> <span class="m">0</span><span class="p">,</span> max <span class="o">=</span> <span class="m">22</span><span class="p">,</span> initial <span class="o">=</span> <span class="m">0</span><span class="p">)</span>
<span class="kr">for</span> <span class="p">(</span>chr <span class="kr">in</span> <span class="m">1</span><span class="o">:</span><span class="m">22</span><span class="p">)</span> <span class="p">{</span>
    setTxtProgressBar<span class="p">(</span>pb<span class="p">,</span> chr<span class="p">)</span>
    <span class="c1"># extract current chromosome</span>
    chr.idx <span class="o">&lt;-</span> <span class="kp">which</span><span class="p">(</span>info_snp<span class="o">$</span>chr <span class="o">==</span> chr<span class="p">)</span>
    df_beta <span class="o">&lt;-</span> info_snp<span class="p">[</span>chr.idx<span class="p">,</span>
                        <span class="kt">c</span><span class="p">(</span><span class="s">&quot;beta&quot;</span><span class="p">,</span> <span class="s">&quot;beta_se&quot;</span><span class="p">,</span> <span class="s">&quot;n_eff&quot;</span><span class="p">)]</span>
    ind.chr <span class="o">&lt;-</span> info_snp<span class="o">$</span><span class="sb">`_NUM_ID_`</span><span class="p">[</span>chr.idx<span class="p">]</span>
    <span class="c1"># calculate LD</span>
    corr0 <span class="o">&lt;-</span> snp_cor<span class="p">(</span>
        genotype<span class="p">,</span>
        ind.col <span class="o">=</span> ind.chr<span class="p">,</span>
        ncores <span class="o">=</span> NCORES<span class="p">,</span>
        infos.pos <span class="o">=</span> POS2<span class="p">[</span>ind.chr<span class="p">],</span>
        size <span class="o">=</span> <span class="m">3</span> <span class="o">/</span> <span class="m">1000</span>
    <span class="p">)</span>
    corr <span class="o">&lt;-</span> bigsparser<span class="o">::</span>as_SFBM<span class="p">(</span>as<span class="p">(</span>corr0<span class="p">,</span> <span class="s">&quot;dgCMatrix&quot;</span><span class="p">))</span>
    <span class="c1"># Perform LDSC analysis to get h2 estimate</span>
    ldsc <span class="o">&lt;-</span> snp_ldsc2<span class="p">(</span>corr0<span class="p">,</span> df_beta<span class="p">)</span>
    h2_est <span class="o">&lt;-</span> ldsc<span class="p">[[</span><span class="s">&quot;h2&quot;</span><span class="p">]]</span>
    <span class="c1"># Get adjusted beta from the auto model</span>
    multi_auto <span class="o">&lt;-</span> snp_ldpred2_auto<span class="p">(</span>
        corr<span class="p">,</span>
        df_beta<span class="p">,</span>
        h2_init <span class="o">=</span> h2_est<span class="p">,</span>
        vec_p_init <span class="o">=</span> seq_log<span class="p">(</span><span class="m">1e-4</span><span class="p">,</span> <span class="m">0.9</span><span class="p">,</span> length.out <span class="o">=</span> NCORES<span class="p">),</span>
        ncores <span class="o">=</span> NCORES
    <span class="p">)</span>
    beta_auto <span class="o">&lt;-</span> <span class="kp">sapply</span><span class="p">(</span>multi_auto<span class="p">,</span> <span class="kr">function</span><span class="p">(</span>auto<span class="p">)</span>
        auto<span class="o">$</span>beta_est<span class="p">)</span>
    pred_auto <span class="o">&lt;-</span>
        big_prodMat<span class="p">(</span>genotype<span class="p">,</span>
                    beta_auto<span class="p">,</span>
                    ind.row <span class="o">=</span> ind.test<span class="p">,</span>
                    ind.col <span class="o">=</span> ind.chr<span class="p">)</span>
    <span class="c1"># scale the PRS generated from AUTO</span>
    pred_scaled <span class="o">&lt;-</span> <span class="kp">apply</span><span class="p">(</span>pred_auto<span class="p">,</span> <span class="m">2</span><span class="p">,</span> sd<span class="p">)</span>
    final_beta_auto <span class="o">&lt;-</span>
        <span class="kp">rowMeans</span><span class="p">(</span>beta_auto<span class="p">[,</span>
                           <span class="kp">abs</span><span class="p">(</span>pred_scaled <span class="o">-</span>
                                   median<span class="p">(</span>pred_scaled<span class="p">))</span> <span class="o">&lt;</span>
                               <span class="m">3</span> <span class="o">*</span> mad<span class="p">(</span>pred_scaled<span class="p">)])</span>
    pred_auto <span class="o">&lt;-</span>
        big_prodVec<span class="p">(</span>genotype<span class="p">,</span>
                    final_beta_auto<span class="p">,</span>
                    ind.row <span class="o">=</span> ind.test<span class="p">,</span>
                    ind.col <span class="o">=</span> ind.chr<span class="p">)</span>
    <span class="c1"># add up the calculated auto PRS</span>
    y<span class="p">[,</span> auto <span class="o">:=</span> auto <span class="o">+</span> pred_auto<span class="p">]</span>
<span class="p">}</span>
<span class="kp">print</span><span class="p">(</span><span class="s">&quot;Completed&quot;</span><span class="p">)</span>
</pre></div></div>
<input name="__tabs_5" type="radio" id="__tab_5_3" />
<label for="__tab_5_3">all model</label>
<div class="superfences-content"><div class="codehilite"><pre><span></span><span class="c1"># Get maximum amount of cores</span>
NCORES <span class="o">&lt;-</span> nb_cores<span class="p">()</span>
<span class="c1"># Prepare data for grid model</span>
p_seq <span class="o">&lt;-</span> <span class="kp">signif</span><span class="p">(</span>seq_log<span class="p">(</span><span class="m">1e-4</span><span class="p">,</span> <span class="m">1</span><span class="p">,</span> length.out <span class="o">=</span> <span class="m">17</span><span class="p">),</span> <span class="m">2</span><span class="p">)</span>
<span class="c1"># Start doing analysis on each chromosome</span>
y<span class="p">[,</span> Inf.est <span class="o">:=</span> <span class="m">0</span><span class="p">]</span>
y<span class="p">[,</span> auto <span class="o">:=</span> <span class="m">0</span><span class="p">]</span>
y<span class="p">[,</span> grid <span class="o">:=</span> <span class="m">0</span><span class="p">]</span>
<span class="c1"># add progress bar</span>
pb <span class="o">=</span> txtProgressBar<span class="p">(</span>min <span class="o">=</span> <span class="m">0</span><span class="p">,</span> max <span class="o">=</span> <span class="m">22</span><span class="p">,</span> initial <span class="o">=</span> <span class="m">0</span><span class="p">)</span>
<span class="kr">for</span> <span class="p">(</span>chr <span class="kr">in</span> <span class="m">1</span><span class="o">:</span><span class="m">22</span><span class="p">)</span> <span class="p">{</span>
    setTxtProgressBar<span class="p">(</span>pb<span class="p">,</span> chr<span class="p">)</span>
    <span class="c1"># Extract current chromosome</span>
    chr.idx <span class="o">&lt;-</span> <span class="kp">which</span><span class="p">(</span>info_snp<span class="o">$</span>chr <span class="o">==</span> chr<span class="p">)</span>
    df_beta <span class="o">&lt;-</span> info_snp<span class="p">[</span>chr.idx<span class="p">,</span>
                        <span class="kt">c</span><span class="p">(</span><span class="s">&quot;beta&quot;</span><span class="p">,</span> <span class="s">&quot;beta_se&quot;</span><span class="p">,</span> <span class="s">&quot;n_eff&quot;</span><span class="p">)]</span>
    ind.chr <span class="o">&lt;-</span> info_snp<span class="o">$</span><span class="sb">`_NUM_ID_`</span><span class="p">[</span>chr.idx<span class="p">]</span>
    <span class="c1"># Calculate LD</span>
    corr0 <span class="o">&lt;-</span> snp_cor<span class="p">(</span>
        genotype<span class="p">,</span>
        ind.col <span class="o">=</span> ind.chr<span class="p">,</span>
        ncores <span class="o">=</span> NCORES<span class="p">,</span>
        infos.pos <span class="o">=</span> POS2<span class="p">[</span>ind.chr<span class="p">],</span>
        size <span class="o">=</span> <span class="m">3</span> <span class="o">/</span> <span class="m">1000</span>
    <span class="p">)</span>
    corr <span class="o">&lt;-</span> bigsparser<span class="o">::</span>as_SFBM<span class="p">(</span>as<span class="p">(</span>corr0<span class="p">,</span> <span class="s">&quot;dgCMatrix&quot;</span><span class="p">))</span>
    <span class="c1"># Perform LDSC analysis</span>
    ldsc <span class="o">&lt;-</span> snp_ldsc2<span class="p">(</span>corr0<span class="p">,</span> df_beta<span class="p">)</span>
    h2_est <span class="o">&lt;-</span> ldsc<span class="p">[[</span><span class="s">&quot;h2&quot;</span><span class="p">]]</span>
    h2_seq <span class="o">&lt;-</span> <span class="kp">round</span><span class="p">(</span>h2_est <span class="o">*</span> <span class="kt">c</span><span class="p">(</span><span class="m">0.7</span><span class="p">,</span> <span class="m">1</span><span class="p">,</span> <span class="m">1.4</span><span class="p">),</span> <span class="m">4</span><span class="p">)</span>
    grid.param <span class="o">&lt;-</span>
        <span class="kp">expand.grid</span><span class="p">(</span>p <span class="o">=</span> p_seq<span class="p">,</span>
                    h2 <span class="o">=</span> h2_seq<span class="p">,</span>
                    sparse <span class="o">=</span> <span class="kt">c</span><span class="p">(</span><span class="kc">FALSE</span><span class="p">,</span> <span class="kc">TRUE</span><span class="p">))</span>
    <span class="c1"># Get adjusted beta from infinitesimal model</span>
    beta_inf <span class="o">&lt;-</span> snp_ldpred2_inf<span class="p">(</span>corr<span class="p">,</span> df_beta<span class="p">,</span> h2 <span class="o">=</span> h2_est<span class="p">)</span>
    <span class="c1"># Get adjusted beta from grid model</span>
    beta_grid <span class="o">&lt;-</span>
        snp_ldpred2_grid<span class="p">(</span>corr<span class="p">,</span> df_beta<span class="p">,</span> grid.param<span class="p">,</span> ncores <span class="o">=</span> NCORES<span class="p">)</span>
    <span class="c1"># Get adjusted beta from the auto model</span>
    multi_auto <span class="o">&lt;-</span> snp_ldpred2_auto<span class="p">(</span>
        corr<span class="p">,</span>
        df_beta<span class="p">,</span>
        h2_init <span class="o">=</span> h2_est<span class="p">,</span>
        vec_p_init <span class="o">=</span> seq_log<span class="p">(</span><span class="m">1e-4</span><span class="p">,</span> <span class="m">0.9</span><span class="p">,</span> length.out <span class="o">=</span> NCORES<span class="p">),</span>
        ncores <span class="o">=</span> NCORES
    <span class="p">)</span>
    beta_auto <span class="o">&lt;-</span> <span class="kp">sapply</span><span class="p">(</span>multi_auto<span class="p">,</span> <span class="kr">function</span><span class="p">(</span>auto<span class="p">)</span>
        auto<span class="o">$</span>beta_est<span class="p">)</span>
    pred_auto <span class="o">&lt;-</span>
        big_prodMat<span class="p">(</span>genotype<span class="p">,</span>
                    beta_auto<span class="p">,</span>
                    ind.row <span class="o">=</span> ind.test<span class="p">,</span>
                    ind.col <span class="o">=</span> ind.chr<span class="p">)</span>
    <span class="c1"># scale the PRS generated from AUTO</span>
    pred_scaled <span class="o">&lt;-</span> <span class="kp">apply</span><span class="p">(</span>pred_auto<span class="p">,</span> <span class="m">2</span><span class="p">,</span> sd<span class="p">)</span>
    final_beta_auto <span class="o">&lt;-</span>
        <span class="kp">rowMeans</span><span class="p">(</span>beta_auto<span class="p">[,</span>
                           <span class="kp">abs</span><span class="p">(</span>pred_scaled <span class="o">-</span>
                                   median<span class="p">(</span>pred_scaled<span class="p">))</span> <span class="o">&lt;</span>
                               <span class="m">3</span> <span class="o">*</span> mad<span class="p">(</span>pred_scaled<span class="p">)])</span>
    <span class="c1"># Get the auto PRS</span>
    pred_auto <span class="o">&lt;-</span>
        big_prodVec<span class="p">(</span>genotype<span class="p">,</span>
                    final_beta_auto<span class="p">,</span>
                    ind.row <span class="o">=</span> ind.test<span class="p">,</span>
                    ind.col <span class="o">=</span> ind.chr<span class="p">)</span>
    <span class="c1"># Get infinitesimal PRS</span>
    pred_inf <span class="o">&lt;-</span>
        big_prodVec<span class="p">(</span>genotype<span class="p">,</span>
                    beta_inf<span class="p">,</span>
                    ind.row <span class="o">=</span> ind.test<span class="p">,</span>
                    ind.col <span class="o">=</span> ind.chr<span class="p">)</span>
    <span class="c1"># Get the grid PRSs</span>
    pred_grid <span class="o">&lt;-</span> big_prodMat<span class="p">(</span>genotype<span class="p">,</span> beta_grid<span class="p">,</span> ind.col <span class="o">=</span> ind.chr<span class="p">)</span>
    <span class="c1"># add up the calculated PRS</span>
    y<span class="p">[,</span> Inf.est <span class="o">:=</span> Inf.est <span class="o">+</span> pred_inf<span class="p">]</span>
    <span class="c1"># for grid, we want to retain the &quot;best&quot; prs only</span>
    indep <span class="o">&lt;-</span> copy<span class="p">(</span>y<span class="p">)</span>
    max.grid.r2 <span class="o">&lt;-</span> <span class="m">0</span>
    max.grid.id <span class="o">&lt;-</span> <span class="m">0</span>
    <span class="kr">for</span> <span class="p">(</span>i <span class="kr">in</span> <span class="m">1</span><span class="o">:</span><span class="kp">ncol</span><span class="p">(</span>pred_grid<span class="p">))</span> <span class="p">{</span>
        indep<span class="p">[,</span> PRS <span class="o">:=</span> pred_grid<span class="p">[,</span> i<span class="p">]]</span>
        model <span class="o">&lt;-</span> lm<span class="p">(</span>reg.formula<span class="p">,</span> data <span class="o">=</span> indep<span class="p">)</span> <span class="o">%&gt;%</span>
            <span class="kp">summary</span>
        prs.r2 <span class="o">&lt;-</span> model<span class="o">$</span>r.squared <span class="o">-</span> null.r2
        <span class="kr">if</span> <span class="p">(</span>max.grid.r2 <span class="o">&lt;</span> prs.r2<span class="p">)</span> <span class="p">{</span>
            max.grid.r2 <span class="o">&lt;-</span> prs.r2
            max.grid.id <span class="o">&lt;-</span> i
        <span class="p">}</span>
    <span class="p">}</span>
    <span class="c1"># add up the calculated grid PRS</span>
    y<span class="p">[,</span> grid <span class="o">:=</span> grid <span class="o">+</span> pred_grid<span class="p">[,</span> max.grid.id<span class="p">]]</span>
    <span class="c1"># add up the calculated auto PRS</span>
    y<span class="p">[,</span> auto <span class="o">:=</span> auto <span class="o">+</span> pred_auto<span class="p">]</span>
<span class="p">}</span>
<span class="kp">print</span><span class="p">(</span><span class="s">&quot;Completed&quot;</span><span class="p">)</span>
</pre></div></div>
</div>
<h1 id="5-get-the-final-performance-of-the-ldpred-models">5. Get the final performance of the LDpred models<a class="headerlink" href="#5-get-the-final-performance-of-the-ldpred-models" title="Permanent link">&para;</a></h1>
<div class="superfences-tabs">
<input name="__tabs_6" type="radio" id="__tab_6_0" checked="checked" />
<label for="__tab_6_0">infinitesimal model</label>
<div class="superfences-content"><div class="codehilite"><pre><span></span><span class="c1"># there are 3 models: inf, auto and grid</span>
inf.model <span class="o">&lt;-</span> <span class="kp">paste</span><span class="p">(</span><span class="s">&quot;PC&quot;</span><span class="p">,</span> <span class="m">1</span><span class="o">:</span><span class="m">6</span><span class="p">,</span> sep <span class="o">=</span> <span class="s">&quot;&quot;</span><span class="p">,</span> collapse <span class="o">=</span> <span class="s">&quot;+&quot;</span><span class="p">)</span> <span class="o">%&gt;%</span>
    <span class="kp">paste0</span><span class="p">(</span><span class="s">&quot;Height~Inf.est+Sex+&quot;</span><span class="p">,</span> <span class="m">.</span><span class="p">)</span> <span class="o">%&gt;%</span>
    as.formula <span class="o">%&gt;%</span>
    lm<span class="p">(</span><span class="m">.</span><span class="p">,</span> data <span class="o">=</span> y<span class="p">)</span> <span class="o">%&gt;%</span>
    <span class="kp">summary</span>

result <span class="o">&lt;-</span> data.table<span class="p">(</span>
    infinitesimal <span class="o">=</span> inf.model<span class="o">$</span>r.squared <span class="o">-</span> null.r2<span class="p">,</span>
    null <span class="o">=</span> null.r2
<span class="p">)</span>
<span class="kp">print</span><span class="p">(</span>result<span class="p">)</span>
</pre></div></div>
<input name="__tabs_6" type="radio" id="__tab_6_1" />
<label for="__tab_6_1">grid model</label>
<div class="superfences-content"><div class="codehilite"><pre><span></span>grid.model <span class="o">&lt;-</span> <span class="kp">paste</span><span class="p">(</span><span class="s">&quot;PC&quot;</span><span class="p">,</span> <span class="m">1</span><span class="o">:</span><span class="m">6</span><span class="p">,</span> sep <span class="o">=</span> <span class="s">&quot;&quot;</span><span class="p">,</span> collapse <span class="o">=</span> <span class="s">&quot;+&quot;</span><span class="p">)</span> <span class="o">%&gt;%</span>
    <span class="kp">paste0</span><span class="p">(</span><span class="s">&quot;Height~grid+Sex+&quot;</span><span class="p">,</span> <span class="m">.</span><span class="p">)</span> <span class="o">%&gt;%</span>
    as.formula <span class="o">%&gt;%</span>
    lm<span class="p">(</span><span class="m">.</span><span class="p">,</span> data <span class="o">=</span> y<span class="p">)</span> <span class="o">%&gt;%</span>
    <span class="kp">summary</span>

result <span class="o">&lt;-</span> data.table<span class="p">(</span>
    grid <span class="o">=</span> grid.model<span class="o">$</span>r.squared <span class="o">-</span> null.r2<span class="p">,</span>
    null <span class="o">=</span> null.r2
<span class="p">)</span>
<span class="kp">print</span><span class="p">(</span>result<span class="p">)</span>
</pre></div></div>
<input name="__tabs_6" type="radio" id="__tab_6_2" />
<label for="__tab_6_2">auto model</label>
<div class="superfences-content"><div class="codehilite"><pre><span></span>auto.model <span class="o">&lt;-</span> <span class="kp">paste</span><span class="p">(</span><span class="s">&quot;PC&quot;</span><span class="p">,</span> <span class="m">1</span><span class="o">:</span><span class="m">6</span><span class="p">,</span> sep <span class="o">=</span> <span class="s">&quot;&quot;</span><span class="p">,</span> collapse <span class="o">=</span> <span class="s">&quot;+&quot;</span><span class="p">)</span> <span class="o">%&gt;%</span>
    <span class="kp">paste0</span><span class="p">(</span><span class="s">&quot;Height~auto+Sex+&quot;</span><span class="p">,</span> <span class="m">.</span><span class="p">)</span> <span class="o">%&gt;%</span>
    as.formula <span class="o">%&gt;%</span>
    lm<span class="p">(</span><span class="m">.</span><span class="p">,</span> data <span class="o">=</span> y<span class="p">)</span> <span class="o">%&gt;%</span>
    <span class="kp">summary</span>

result <span class="o">&lt;-</span> data.table<span class="p">(</span>
    infinitesimal <span class="o">=</span> inf.model<span class="o">$</span>r.squared <span class="o">-</span> null.r2<span class="p">,</span>
    grid <span class="o">=</span> grid.model<span class="o">$</span>r.squared <span class="o">-</span> null.r2<span class="p">,</span>
    auto <span class="o">=</span> auto.model<span class="o">$</span>r.squared <span class="o">-</span> null.r2<span class="p">,</span>
    null <span class="o">=</span> null.r2
<span class="p">)</span>
<span class="kp">print</span><span class="p">(</span>result<span class="p">)</span>
</pre></div></div>
<input name="__tabs_6" type="radio" id="__tab_6_3" />
<label for="__tab_6_3">all model</label>
<div class="superfences-content"><div class="codehilite"><pre><span></span><span class="c1"># there are 3 models: inf, auto and grid</span>
inf.model <span class="o">&lt;-</span> <span class="kp">paste</span><span class="p">(</span><span class="s">&quot;PC&quot;</span><span class="p">,</span> <span class="m">1</span><span class="o">:</span><span class="m">6</span><span class="p">,</span> sep <span class="o">=</span> <span class="s">&quot;&quot;</span><span class="p">,</span> collapse <span class="o">=</span> <span class="s">&quot;+&quot;</span><span class="p">)</span> <span class="o">%&gt;%</span>
    <span class="kp">paste0</span><span class="p">(</span><span class="s">&quot;Height~Inf.est+Sex+&quot;</span><span class="p">,</span> <span class="m">.</span><span class="p">)</span> <span class="o">%&gt;%</span>
    as.formula <span class="o">%&gt;%</span>
    lm<span class="p">(</span><span class="m">.</span><span class="p">,</span> data <span class="o">=</span> y<span class="p">)</span> <span class="o">%&gt;%</span>
    <span class="kp">summary</span>
auto.model <span class="o">&lt;-</span> <span class="kp">paste</span><span class="p">(</span><span class="s">&quot;PC&quot;</span><span class="p">,</span> <span class="m">1</span><span class="o">:</span><span class="m">6</span><span class="p">,</span> sep <span class="o">=</span> <span class="s">&quot;&quot;</span><span class="p">,</span> collapse <span class="o">=</span> <span class="s">&quot;+&quot;</span><span class="p">)</span> <span class="o">%&gt;%</span>
    <span class="kp">paste0</span><span class="p">(</span><span class="s">&quot;Height~auto+Sex+&quot;</span><span class="p">,</span> <span class="m">.</span><span class="p">)</span> <span class="o">%&gt;%</span>
    as.formula <span class="o">%&gt;%</span>
    lm<span class="p">(</span><span class="m">.</span><span class="p">,</span> data <span class="o">=</span> y<span class="p">)</span> <span class="o">%&gt;%</span>
    <span class="kp">summary</span>
grid.model <span class="o">&lt;-</span> <span class="kp">paste</span><span class="p">(</span><span class="s">&quot;PC&quot;</span><span class="p">,</span> <span class="m">1</span><span class="o">:</span><span class="m">6</span><span class="p">,</span> sep <span class="o">=</span> <span class="s">&quot;&quot;</span><span class="p">,</span> collapse <span class="o">=</span> <span class="s">&quot;+&quot;</span><span class="p">)</span> <span class="o">%&gt;%</span>
    <span class="kp">paste0</span><span class="p">(</span><span class="s">&quot;Height~grid+Sex+&quot;</span><span class="p">,</span> <span class="m">.</span><span class="p">)</span> <span class="o">%&gt;%</span>
    as.formula <span class="o">%&gt;%</span>
    lm<span class="p">(</span><span class="m">.</span><span class="p">,</span> data <span class="o">=</span> y<span class="p">)</span> <span class="o">%&gt;%</span>
    <span class="kp">summary</span>

result <span class="o">&lt;-</span> data.table<span class="p">(</span>
    infinitesimal <span class="o">=</span> inf.model<span class="o">$</span>r.squared <span class="o">-</span> null.r2<span class="p">,</span>
    grid <span class="o">=</span> grid.model<span class="o">$</span>r.squared <span class="o">-</span> null.r2<span class="p">,</span>
    auto <span class="o">=</span> auto.model<span class="o">$</span>r.squared <span class="o">-</span> null.r2<span class="p">,</span>
    null <span class="o">=</span> null.r2
<span class="p">)</span>
<span class="kp">print</span><span class="p">(</span>result<span class="p">)</span>
</pre></div></div>
</div>
<details class="note"><summary>How much phenotypic variation does the PRS from each model explain?</summary><p>Infinitesimal = 0.1717342</p>
<p>Grid Model = 0.002945278</p>
<p>Auto Model = 0.2532782</p>
</details>
              
            </div>
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="../lassosum/" class="btn btn-neutral float-right" title="lassosum">Next <span class="icon icon-circle-arrow-right"></span></a>
      
      
        <a href="../prsice/" class="btn btn-neutral" title="PRSice-2"><span class="icon icon-circle-arrow-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <!-- Copyright etc -->
    
  </div>

  Built with <a href="http://www.mkdocs.org">MkDocs</a> using a <a href="https://github.com/snide/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>.
</footer>
      
        </div>
      </div>

    </section>

  </div>

  <div class="rst-versions" role="note" style="cursor: pointer">
    <span class="rst-current-version" data-toggle="rst-current-version">
      
          <a href="https://github.com/choishingwan/PRS-Tutorial/" class="fa fa-github" style="float: left; color: #fcfcfc"> GitHub</a>
      
      
        <span><a href="../prsice/" style="color: #fcfcfc;">&laquo; Previous</a></span>
      
      
        <span style="margin-left: 15px"><a href="../lassosum/" style="color: #fcfcfc">Next &raquo;</a></span>
      
    </span>
</div>
    <script>var base_url = '..';</script>
    <script src="../js/theme.js" defer></script>
      <script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.0/MathJax.js?config=TeX-MML-AM_CHTML" defer></script>
      <script src="../search/main.js" defer></script>

</body>
</html>
