<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <meta name="author" content="Shing Wan CHOI">
  <link rel="shortcut icon" href="../img/favicon.ico">
  <title>1. QC of Base Data - Basic Tutorial for Polygenic Risk Score Analyses</title>
  <link href='https://fonts.googleapis.com/css?family=Lato:400,700|Roboto+Slab:400,700|Inconsolata:400,700' rel='stylesheet' type='text/css'>

  <link rel="stylesheet" href="../css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../css/theme_extra.css" type="text/css" />
  <link rel="stylesheet" href="//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/styles/github.min.css">
  <link href="../css/extra.css" rel="stylesheet">
  
  <script>
    // Current page data
    var mkdocs_page_name = "1. QC of Base Data";
    var mkdocs_page_input_path = "base.md";
    var mkdocs_page_url = null;
  </script>
  
  <script src="../js/jquery-2.1.1.min.js" defer></script>
  <script src="../js/modernizr-2.8.3.min.js" defer></script>
  <script src="//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/highlight.min.js"></script>
  <script>hljs.initHighlightingOnLoad();</script> 
  
  <script>
      (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
      (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
      m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
      })(window,document,'script','https://www.google-analytics.com/analytics.js','ga');

      ga('create', 'UA-145068952-1', 'auto');
      ga('send', 'pageview');
  </script>
  
</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side stickynav">
      <div class="wy-side-nav-search">
        <a href=".." class="icon icon-home"> Basic Tutorial for Polygenic Risk Score Analyses</a>
        <div role="search">
  <form id ="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" title="Type search term here" />
  </form>
</div>
      </div>

      <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
	<ul class="current">
	  
          
            <li class="toctree-l1">
		
    <a class="" href="..">Overview</a>
	    </li>
          
            <li class="toctree-l1 current">
		
    <a class="current" href="./">1. QC of Base Data</a>
    <ul class="subnav">
            
    <li class="toctree-l2"><a href="#obtaining-the-base-data-file">Obtaining the base data file</a></li>
    

    <li class="toctree-l2"><a href="#reading-the-base-data-file">Reading the base data file</a></li>
    

    <li class="toctree-l2"><a href="#qc-checklist-base-data">QC checklist: Base data</a></li>
    

    <li class="toctree-l2"><a href="#35-heritability-check"># Heritability check</a></li>
    

    <li class="toctree-l2"><a href="#35-effect-allele"># Effect allele</a></li>
    

    <li class="toctree-l2"><a href="#35-file-transfer"># File transfer</a></li>
    

    <li class="toctree-l2"><a href="#35-genome-build"># Genome build</a></li>
    

    <li class="toctree-l2"><a href="#35-standard-gwas-qc"># Standard GWAS QC</a></li>
    

    <li class="toctree-l2"><a href="#35-mismatching-snps"># Mismatching SNPs</a></li>
    

    <li class="toctree-l2"><a href="#35-duplicate-snps"># Duplicate SNPs</a></li>
    

    <li class="toctree-l2"><a href="#35-ambiguous-snps"># Ambiguous SNPs</a></li>
    

    <li class="toctree-l2"><a href="#35-sex-chromosomes"># Sex chromosomes</a></li>
    

    <li class="toctree-l2"><a href="#35-sample-overlap"># Sample overlap</a></li>
    

    <li class="toctree-l2"><a href="#35-relatedness"># Relatedness</a></li>
    

    </ul>
	    </li>
          
            <li class="toctree-l1">
		
    <a class="" href="../target/">2. QC of Target Data</a>
	    </li>
          
            <li class="toctree-l1">
		
    <span class="caption-text">3. Calculating and analysing PRS</span>
    <ul class="subnav">
                <li class="">
                    
    <a class="" href="../plink/">PLINK</a>
                </li>
                <li class="">
                    
    <a class="" href="../prsice/">PRSice-2</a>
                </li>
                <li class="">
                    
    <a class="" href="../ldpred/">LDpred</a>
                </li>
                <li class="">
                    
    <a class="" href="../lassosum/">lassosum</a>
                </li>
    </ul>
	    </li>
          
            <li class="toctree-l1">
		
    <a class="" href="../plink_visual/">4. Visualizing PRS Results</a>
	    </li>
          
        </ul>
      </div>
      &nbsp;
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" role="navigation" aria-label="top navigation">
        <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
        <a href="..">Basic Tutorial for Polygenic Risk Score Analyses</a>
      </nav>

      
      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href="..">Docs</a> &raquo;</li>
    
      
    
    <li>1. QC of Base Data</li>
    <li class="wy-breadcrumbs-aside">
      
        <a href="https://github.com/choishingwan/PRS-Tutorial/edit/master/docs/base.md"
          class="icon icon-github"> Edit on GitHub</a>
      
    </li>
  </ul>
  <hr/>
</div>
          <div role="main">
            <div class="section">
              
                <h1 id="obtaining-the-base-data-file">Obtaining the base data file<a class="headerlink" href="#obtaining-the-base-data-file" title="Permanent link">&para;</a></h1>
<p>The first step in Polygenic Risk Score (PRS) analyses is to generate or obtain the base data (GWAS summary statistics). Ideally these will correspond to the most powerful GWAS results available on the phenotype under study. In this example, we will use GWAS on simulated height. You can download the summary statistic file <a href="https://drive.google.com/file/d/1RWjk49QNZj9zvJHc9X_wyZ51fdy6xQjv/view?usp=sharing">here</a> </p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Due to limitation to bandwidth, we are currently using google drive to host the files, which doesn't allow the use of wget or curl to download the file. Please download the files manually.</p>
</div>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>If you download the summary statistics on a MAC machine, the gz file will be decompressed automatically, resulting in a <strong>Height.gwas.txt</strong> file instead. </p>
<p>To maintain consistency, we suggest compressing the <strong>Height.gwas.txt</strong> file with 
<div class="codehilite"><pre><span></span>gzip Height.gwas.txt
</pre></div>
before starting the tutorial</p>
</div>
<h1 id="reading-the-base-data-file">Reading the base data file<a class="headerlink" href="#reading-the-base-data-file" title="Permanent link">&para;</a></h1>
<p><strong>Height.gwas.txt.gz</strong> is compressed. To read its content, you can type:</p>
<div class="codehilite"><pre><span></span>gunzip -c Height.gwas.txt.gz <span class="p">|</span> head
</pre></div>

<p>which will display the first 10 lines of the file</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Working with compressed files reduces storage space requirements</p>
</div>
<p>The <strong>Height.gwas.txt.gz</strong> file contains the following columns:</p>
<table>
<thead>
<tr>
<th align="center">CHR</th>
<th align="center">BP</th>
<th align="center">SNP</th>
<th align="center">A1</th>
<th align="center">A2</th>
<th align="center">N</th>
<th align="center">SE</th>
<th align="center">P</th>
<th align="center">OR</th>
<th align="center">INFO</th>
<th align="center">MAF</th>
</tr>
</thead>
<tbody>
<tr>
<td align="center">1</td>
<td align="center">756604</td>
<td align="center">rs3131962</td>
<td align="center">A</td>
<td align="center">G</td>
<td align="center">388028</td>
<td align="center">0.00301666</td>
<td align="center">0.483171</td>
<td align="center">0.997886915712657</td>
<td align="center">0.890557941364774</td>
<td align="center">0.369389592764921</td>
</tr>
<tr>
<td align="center">1</td>
<td align="center">768448</td>
<td align="center">rs12562034</td>
<td align="center">A</td>
<td align="center">G</td>
<td align="center">388028</td>
<td align="center">0.00329472</td>
<td align="center">0.834808</td>
<td align="center">1.00068731609353</td>
<td align="center">0.895893511351165</td>
<td align="center">0.336845754096289</td>
</tr>
<tr>
<td align="center">1</td>
<td align="center">779322</td>
<td align="center">rs4040617</td>
<td align="center">G</td>
<td align="center">A</td>
<td align="center">388028</td>
<td align="center">0.00303344</td>
<td align="center">0.42897</td>
<td align="center">0.997603556067569</td>
<td align="center">0.897508290615237</td>
<td align="center">0.377368010940814</td>
</tr>
</tbody>
</table>
<p>The column headers correspond to the following: </p>
<ul>
<li><strong>CHR</strong>: The chromosome in which the SNP resides</li>
<li><strong>BP</strong>: Chromosomal co-ordinate of the SNP</li>
<li><strong>SNP</strong>: SNP ID, usually in the form of rs-ID</li>
<li><strong>A1</strong>: The effect allele of the SNP</li>
<li><strong>A2</strong>: The non-effect allele of the SNP</li>
<li><strong>N</strong>: Number of samples used to obtain the effect size estimate</li>
<li><strong>SE</strong>: The standard error (SE) of the effect size esimate</li>
<li><strong>P</strong>: The P-value of association between the SNP genotypes and the base phenotype</li>
<li><strong>OR</strong>: The effect size estimate of the SNP, if the outcome is binary/case-control. If the outcome is continuous or treated as continuous then this will usually be BETA</li>
<li><strong>INFO</strong>: The imputation information score</li>
<li><strong>MAF</strong>: The minor allele frequency (MAF) of the SNP</li>
</ul>
<h1 id="qc-checklist-base-data">QC checklist: Base data<a class="headerlink" href="#qc-checklist-base-data" title="Permanent link">&para;</a></h1>
<p>Below we perform QC on these base data according to the 'QC checklist' in our <a href="https://doi.org/10.1101/416545">guide paper</a>, which we recommend that users follow while going through this tutorial and when performing PRS analyses:</p>
<h1 id="35-heritability-check"># Heritability check<a class="headerlink" href="#35-heritability-check" title="Permanent link">&para;</a></h1>
<p>We recommend that PRS analyses are performed on base data with a chip-heritability estimate <span class="arithmatex">\(h_{snp}^{2} &gt; 0.05\)</span>. 
The chip-heritability of a GWAS can be estimated using e.g. LD Score Regression (LDSC). 
Our height GWAS data are simulated to have a chip-heritability much greater than 0.05 and so we can move on to the next QC step. </p>
<h1 id="35-effect-allele"># Effect allele<a class="headerlink" href="#35-effect-allele" title="Permanent link">&para;</a></h1>
<p>It is important to know which allele is the effect allele and which is the non-effect allele for PRS association results to be in the correct direction.</p>
<div class="admonition important">
<p class="admonition-title">Important</p>
<p>Some GWAS results files do not make clear which allele is the effect allele and which is the non-effect allele.
If the incorrect assumption is made in computing the PRS, then the effect of the PRS in the target data will be in the wrong direction.</p>
<p>To avoid misleading conclusions the effect allele from the base (GWAS) data must be known.</p>
</div>
<h1 id="35-file-transfer"># File transfer<a class="headerlink" href="#35-file-transfer" title="Permanent link">&para;</a></h1>
<p>A common problem is that the downloaded base data file can be
corrupted during download, which can cause PRS software to crash 
or to produce errors in results. 
However, a <code>md5sum</code> hash is 
generally included in files so that file integrity can be checked. 
The following command performs this <code>md5sum</code> check: </p>
<div class="superfences-tabs">
<input name="__tabs_1" type="radio" id="__tab_1_0" checked="checked" />
<label for="__tab_1_0">Linux</label>
<div class="superfences-content"><div class="codehilite"><pre><span></span>md5sum Height.gwas.txt.gz
</pre></div></div>
<input name="__tabs_1" type="radio" id="__tab_1_1" />
<label for="__tab_1_1">OS X</label>
<div class="superfences-content"><div class="codehilite"><pre><span></span>md5 Height.gwas.txt.gz
</pre></div></div>
</div>
<p>if the file is intact, then <code>md5sum</code> generates a string of characters, which in this case should be: <code>a2b15fb6a2bbbe7ef49f67959b43b160</code>. 
If a different string is generated, then the file is corrupted.</p>
<h1 id="35-genome-build"># Genome build<a class="headerlink" href="#35-genome-build" title="Permanent link">&para;</a></h1>
<p>The height summary statistic are on the same genome build as the target data that we will be using. 
You must check that your base and target data are on the same genome build, and if they are not then use a tool such as <a href="https://genome.ucsc.edu/cgi-bin/hgLiftOver">LiftOver</a> to make the builds consistent across the data sets.</p>
<h1 id="35-standard-gwas-qc"># Standard GWAS QC<a class="headerlink" href="#35-standard-gwas-qc" title="Permanent link">&para;</a></h1>
<p>As described in the paper, both the base and target data should be subjected to the standard stringent QC steps performed in GWAS. 
If the base data have been obtained as summary statistics from a public source, then the typical QC steps that you will be able to perform on them are to filter the SNPs according to INFO score and MAF. 
SNPs with low minor allele frequency (MAF) or imputation information score (INFO) are more likely to generate false positive results due to their lower statistical power (and higher probability of genotyping errors in the case of low MAF). 
Therefore, SNPs with low MAF and INFO are typically removed before performing downstream analyses.
We recommend removing SNPs with MAF &lt; 1% and INFO &lt; 0.8 (with very large base sample sizes these thresholds could be reduced if sensitivity checks indicate reliable results).
These SNP filters can be achieved using the following code:</p>
<div class="superfences-tabs">
<input name="__tabs_2" type="radio" id="__tab_2_0" checked="checked" />
<label for="__tab_2_0">Using bash</label>
<div class="superfences-content"><div class="codehilite"><pre><span></span>gunzip -c Height.gwas.txt.gz <span class="p">|</span><span class="se">\</span>
awk <span class="s1">&#39;NR==1 || ($11 &gt; 0.01) &amp;&amp; ($10 &gt; 0.8) {print}&#39;</span> <span class="p">|</span><span class="se">\</span>
gzip  &gt; Height.gz
</pre></div></div>
<input name="__tabs_2" type="radio" id="__tab_2_1" />
<label for="__tab_2_1">Using R with data.table</label>
<div class="superfences-content"><div class="codehilite"><pre><span></span><span class="c1"># Alternatively, you can use R, with data.table v1.11.8+</span>
<span class="kn">library</span><span class="p">(</span>data.table<span class="p">)</span>
<span class="c1"># Read in file</span>
dat <span class="o">&lt;-</span> fread<span class="p">(</span><span class="s">&quot;Height.gwas.txt.gz&quot;</span><span class="p">)</span>
<span class="c1"># Filter out SNPs</span>
result <span class="o">&lt;-</span> dat<span class="p">[</span>INFO <span class="o">&gt;</span> <span class="m">0.8</span> <span class="o">&amp;</span> MAF <span class="o">&lt;</span> <span class="m">0.01</span><span class="p">]</span>
<span class="c1"># Output the gz file</span>
fwrite<span class="p">(</span>result<span class="p">,</span> <span class="s">&quot;Height.gz&quot;</span><span class="p">,</span> sep<span class="o">=</span><span class="s">&quot;\t&quot;</span><span class="p">)</span>
</pre></div></div>
</div>
<p>The bash code above does the following:</p>
<ol>
<li>Decompresses and reads the <strong>Height.gwas.txt.gz</strong> file</li>
<li>Prints the header line (<code>NR==1</code>)</li>
<li>Prints any line with MAF above 0.01 (<code>$11</code> because the sixth column of the file contains the MAF information)</li>
<li>Prints any line with INFO above 0.8 (<code>$10</code> because the tenth column of the file contains the INFO information)</li>
<li>Compresses and writes the results to <strong>Height.gz</strong></li>
</ol>
<h1 id="35-mismatching-snps"># Mismatching SNPs<a class="headerlink" href="#35-mismatching-snps" title="Permanent link">&para;</a></h1>
<p>SNPs that have mismatching alleles reported in the base and target data are either resolvable by "strand-flipping" the alleles to their complementary alleles in e.g. the target data, such as for a SNP with A/C in the base data and G/T in the target, or non-resolvable, such as for a SNP with C/G in the base and C/T in the target. 
Most polygenic score software perform strand-flipping automatically for SNPs that are resolvable, and remove non-resolvable mismatching SNPs.</p>
<p>Since we need the target data to know which SNPs have mismatching alleles, we will perform this strand-flipping in the target data.</p>
<h1 id="35-duplicate-snps"># Duplicate SNPs<a class="headerlink" href="#35-duplicate-snps" title="Permanent link">&para;</a></h1>
<p>If an error has occurred in the generation of the base data then there may be duplicated SNPs in the base data file.
Most PRS software do not allow duplicated SNPs in the base data input and thus they should be removed, using a command such as the one below: </p>
<div class="codehilite"><pre><span></span>gunzip -c Height.gz <span class="p">|</span><span class="se">\</span>
awk <span class="s1">&#39;{ print $3}&#39;</span> <span class="p">|</span><span class="se">\</span>
sort <span class="p">|</span><span class="se">\</span>
uniq -d &gt; duplicated.snp
</pre></div>

<p>The above command does the following:</p>
<ol>
<li>Decompresses and reads the <strong>Height.gz</strong> file</li>
<li>Prints out the third column of the file (which contains the SNP ID; change <code>$3</code> to another number if the SNP ID is located in another column, e.g. <code>$1</code> if the SNP ID is located on the first column)</li>
<li>Sort the SNP IDs. This will put duplicated SNP IDs next to each other</li>
<li>Print out any duplicated SNP IDs using the uniq command and print them to the <em>duplicated.snp</em> file</li>
</ol>
<details class="note"><summary>How many duplicated SNPs are there?</summary><p>There are a total of <code>2</code> duplicated SNPs</p>
</details>
<p>Duplicated SNPs can then be removed using the <code>grep</code> command:
<div class="codehilite"><pre><span></span>gunzip -c Height.gz  <span class="p">|</span><span class="se">\</span>
grep -vf duplicated.snp <span class="p">|</span><span class="se">\</span>
gzip - &gt; Height.nodup.gz
</pre></div></p>
<p>The above command does the following:</p>
<ol>
<li>Decompresses and reads the <strong>Height.gz</strong> file</li>
<li>From the file, remove (<code>-v</code>) any lines contains string within the <strong>duplicated.snp</strong> file (<code>-f</code>)</li>
<li>Compresses and writes the results to <strong>Height.nodup.gz</strong></li>
</ol>
<h1 id="35-ambiguous-snps"># Ambiguous SNPs<a class="headerlink" href="#35-ambiguous-snps" title="Permanent link">&para;</a></h1>
<p>If the base and target data were generated using different genotyping chips and the chromosome strand (+/-) that was used for either is unknown, then it is not possible to pair-up the alleles of ambiguous SNPs (i.e. those with complementary alleles, either C/G or A/T SNPs) across the data sets, because it will be unknown whether the base and target data are referring to the same allele or not. While allele frequencies could be used to infer which alleles are on the same strand, the accuracy of this could be low for SNPs with MAF close to 50% or when the base and target data are from different populations. Therefore, we recommend removing all ambiguous SNPs to avoid introducing this potential source of systematic error.</p>
<p>Ambiguous SNPs can be removed in the base data and then there will be no such SNPs in the subsequent analyses, since analyses are performed only on SNPs that overlap between the base and target data.</p>
<p>Nonambiguous SNPs can be retained using the following:
<div class="codehilite"><pre><span></span>gunzip -c Height.nodup.gz <span class="p">|</span><span class="se">\</span>
awk <span class="s1">&#39;!( ($4==&quot;A&quot; &amp;&amp; $5==&quot;T&quot;) || \</span>
<span class="s1">        ($4==&quot;T&quot; &amp;&amp; $5==&quot;A&quot;) || \</span>
<span class="s1">        ($4==&quot;G&quot; &amp;&amp; $5==&quot;C&quot;) || \</span>
<span class="s1">        ($4==&quot;C&quot; &amp;&amp; $5==&quot;G&quot;)) {print}&#39;</span> <span class="p">|</span><span class="se">\</span>
    gzip &gt; Height.QC.gz
</pre></div></p>
<details class="note"><summary>How many non-ambiguous SNPs were there?</summary><p>There are <code>499,617</code> non-ambiguous SNPs</p>
</details>
<h1 id="35-sex-chromosomes"># Sex chromosomes<a class="headerlink" href="#35-sex-chromosomes" title="Permanent link">&para;</a></h1>
<p>Sometimes sample mislabelling can occur, which may lead to invalid results. One indication of a mislabelled sample is a difference between reported sex and that indicated by the sex chromosomes. While this may be due to a difference in sex and gender identity, it could also reflect mislabeling of samples or misreporting and, thus, individuals in which there is a mismatch between biological and reported sex are typically removed. See the Target Data section in which a sex-check is performed.</p>
<h1 id="35-sample-overlap"># Sample overlap<a class="headerlink" href="#35-sample-overlap" title="Permanent link">&para;</a></h1>
<p>Since the target data were simulated there are no overlapping samples between the base and target data here (see the relevant section of <a href="https://doi.org/10.1101/416545">the paper</a> for discussion of the importance of avoiding sample overlap). </p>
<h1 id="35-relatedness"># Relatedness<a class="headerlink" href="#35-relatedness" title="Permanent link">&para;</a></h1>
<p>Closely related individuals within and between the base and the target data may lead to overfitted results, limiting the generalizability of the results (see the relevant sections of <a href="https://doi.org/10.1101/416545">the paper</a>). Relatedness within the target data is tested in the Target Data section.</p>
<p>The <strong>Height.QC.gz</strong> base data are now ready for using in downstream analyses.</p>
              
            </div>
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="../target/" class="btn btn-neutral float-right" title="2. QC of Target Data">Next <span class="icon icon-circle-arrow-right"></span></a>
      
      
        <a href=".." class="btn btn-neutral" title="Overview"><span class="icon icon-circle-arrow-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <!-- Copyright etc -->
    
  </div>

  Built with <a href="http://www.mkdocs.org">MkDocs</a> using a <a href="https://github.com/snide/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>.
</footer>
      
        </div>
      </div>

    </section>

  </div>

  <div class="rst-versions" role="note" style="cursor: pointer">
    <span class="rst-current-version" data-toggle="rst-current-version">
      
          <a href="https://github.com/choishingwan/PRS-Tutorial/" class="fa fa-github" style="float: left; color: #fcfcfc"> GitHub</a>
      
      
        <span><a href=".." style="color: #fcfcfc;">&laquo; Previous</a></span>
      
      
        <span style="margin-left: 15px"><a href="../target/" style="color: #fcfcfc">Next &raquo;</a></span>
      
    </span>
</div>
    <script>var base_url = '..';</script>
    <script src="../js/theme.js" defer></script>
      <script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.0/MathJax.js?config=TeX-MML-AM_CHTML" defer></script>
      <script src="../search/main.js" defer></script>

</body>
</html>
